<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>


  <!--radix_placeholder_meta_tags-->
  <title>Class 10: Introduction to heatmaps</title>



  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2023-12-11"/>
  <meta property="article:created" itemprop="dateCreated" content="2023-12-11"/>
  <meta name="article:author" content="Kristen Wells"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="Class 10: Introduction to heatmaps"/>
  <meta property="og:type" content="article"/>
  <meta property="og:locale" content="en_US"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="Class 10: Introduction to heatmaps"/>

  <!--/radix_placeholder_meta_tags-->
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","author","output","draft","date"]}},"value":[{"type":"character","attributes":{},"value":["Class 10: Introduction to heatmaps"]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","url","affiliation","orcid_id"]}},"value":[{"type":"character","attributes":{},"value":["Kristen Wells"]},{"type":"character","attributes":{},"value":["https://github.com/kwells4"]},{"type":"character","attributes":{},"value":["RNA Bioscience Iniative, Barbara Davis Diabetes Center"]},{"type":"character","attributes":{},"value":["0000-0002-7466-8164"]}]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc"]}},"value":[{"type":"logical","attributes":{},"value":[false]},{"type":"logical","attributes":{},"value":[true]}]}]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["12-11-2023"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["class-10-heatmap_files/anchor-4.2.2/anchor.min.js","class-10-heatmap_files/bowser-1.9.3/bowser.min.js","class-10-heatmap_files/distill-2.2.21/template.v2.js","class-10-heatmap_files/figure-html5/unnamed-chunk-10-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-13-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-14-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-15-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-18-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-20-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-22-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-26-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-27-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-29-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-31-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-33-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-34-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-35-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-38-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-39-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-4-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-42-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-43-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-44-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-47-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-5-1.png","class-10-heatmap_files/figure-html5/unnamed-chunk-7-1.png","class-10-heatmap_files/header-attrs-2.14/header-attrs.js","class-10-heatmap_files/header-attrs-2.25/header-attrs.js","class-10-heatmap_files/jquery-3.6.0/jquery-3.6.0.js","class-10-heatmap_files/jquery-3.6.0/jquery-3.6.0.min.js","class-10-heatmap_files/jquery-3.6.0/jquery-3.6.0.min.map","class-10-heatmap_files/popper-2.6.0/popper.min.js","class-10-heatmap_files/tippy-6.2.7/tippy-bundle.umd.min.js","class-10-heatmap_files/tippy-6.2.7/tippy-light-border.css","class-10-heatmap_files/tippy-6.2.7/tippy.css","class-10-heatmap_files/tippy-6.2.7/tippy.umd.min.js","class-10-heatmap_files/webcomponents-2.0.0/webcomponents.js"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
    font-size: 100%;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme
    $('code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        const citeChild = $(this).children()[0]
        // Do not process if @xyz has been used without escaping and without bibliography activated
        // https://github.com/rstudio/distill/issues/466
        if (citeChild === undefined) return true

        if (citeChild.nodeName == "D-FOOTNOTE") {
          var fn = citeChild
          $(this).html(fn.shadowRoot.querySelector("sup"))
          $(this).id = fn.id
          fn.remove()
        }
        var refs = $(this).attr('data-cites').split(" ");
        var refHtml = refs.map(function(ref) {
          // Could use CSS.escape too here, we insure backward compatibility in navigator
          return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
        }).join("\n");
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // fix footnotes in tables (#411)
      // replacing broken distill.pub feature
      $('table d-footnote').each(function() {
        // we replace internal showAtNode methode which is triggered when hovering a footnote
        this.hoverBox.showAtNode = function(node) {
          // ported from https://github.com/distillpub/template/pull/105/files
          calcOffset = function(elem) {
              let x = elem.offsetLeft;
              let y = elem.offsetTop;
              // Traverse upwards until an `absolute` element is found or `elem`
              // becomes null.
              while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                  x += elem.offsetLeft;
                  y += elem.offsetTop;
              }

              return { left: x, top: y };
          }
          // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
          const bbox = node.getBoundingClientRect();
          const offset = calcOffset(node);
          this.show([offset.left + bbox.width, offset.top + bbox.height]);
        }
      })

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      // ignore leaflet img layers (#106)
      figures = figures.filter(':not(img[class*="leaflet"])')
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="class-10-heatmap_files/header-attrs-2.25/header-attrs.js"></script>
  <script src="class-10-heatmap_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="class-10-heatmap_files/popper-2.6.0/popper.min.js"></script>
  <link href="class-10-heatmap_files/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="class-10-heatmap_files/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="class-10-heatmap_files/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="class-10-heatmap_files/anchor-4.2.2/anchor.min.js"></script>
  <script src="class-10-heatmap_files/bowser-1.9.3/bowser.min.js"></script>
  <script src="class-10-heatmap_files/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="class-10-heatmap_files/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"Class 10: Introduction to heatmaps","authors":[{"author":"Kristen Wells","authorURL":"https://github.com/kwells4","affiliation":"RNA Bioscience Iniative, Barbara Davis Diabetes Center","affiliationURL":"#","orcidID":"0000-0002-7466-8164"}],"publishedDate":"2023-12-11T00:00:00.000-07:00","citationText":"Wells, 2023"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Class 10: Introduction to heatmaps</h1>
<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->

</div>

<div class="d-byline">
  Kristen Wells <a href="https://github.com/kwells4" class="uri">https://github.com/kwells4</a> (RNA Bioscience Iniative, Barbara Davis Diabetes Center)
  
<br/>12-11-2023
</div>

<div class="d-article">
<div class="d-contents d-contents-float">
<nav class="l-text toc figcaption" id="TOC">
<h3>Contents</h3>
<ul>
<li><a href="#goals-for-this-class" id="toc-goals-for-this-class">Goals for this class</a></li>
<li><a href="#load-packages" id="toc-load-packages">Load packages</a></li>
<li><a href="#download-files" id="toc-download-files">Download files</a></li>
<li><a href="#making-a-heatmap" id="toc-making-a-heatmap">Making a heatmap</a>
<ul>
<li><a href="#cleaning-up-by-normalizing-values" id="toc-cleaning-up-by-normalizing-values">Cleaning up by normalizing values</a></li>
<li><a href="#cleaning-up-by-scaling-values" id="toc-cleaning-up-by-scaling-values">Cleaning up by scaling values</a></li>
<li><a href="#changing-the-color-palette" id="toc-changing-the-color-palette">Changing the color palette</a></li>
<li><a href="#adding-annotations" id="toc-adding-annotations">Adding annotations</a></li>
<li><a href="#clustering" id="toc-clustering">Clustering</a></li>
<li><a href="#remove-column-and-row-names" id="toc-remove-column-and-row-names">Remove column and row names</a></li>
<li><a href="#other-aesthetics" id="toc-other-aesthetics">Other aesthetics</a></li>
</ul></li>
<li><a href="#acknowldgements-and-additional-references" id="toc-acknowldgements-and-additional-references">Acknowldgements and additional references</a></li>
<li><a href="#im-a-biologist-why-should-i-care" id="toc-im-a-biologist-why-should-i-care">I’m a biologist, why should I care?</a></li>
</ul>
</nav>
</div>
<p><em>The Rmarkdown for this class is <a href="https://raw.githubusercontent.com/rnabioco/bmsc-7810-pbda/main/_posts/2023-12-12-class-10-heatmap/class-10-heatmap.Rmd">on github</a></em></p>
<h2 id="goals-for-this-class">Goals for this class</h2>
<ul>
<li>Learn how to make a heat map using <code>pheatmap</code></li>
<li>Understand how data is generally processed before making a heat map, understand what interpretations can be made given the processing.</li>
<li>Learn how to change the aesthetics of a heat map</li>
<li>Learn how to visualize and access clustering information from the heat map</li>
</ul>
<h2 id="load-packages">Load packages</h2>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://tidyverse.tidyverse.org'>tidyverse</a></span><span class='op'>)</span></span>
<span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>pheatmap</span><span class='op'>)</span> <span class='co'># install.packages("pheatmap")</span></span>
<span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://sjmgarnier.github.io/viridis/'>viridis</a></span><span class='op'>)</span> <span class='co'># install.packages("viridis")</span></span>
<span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>MetBrewer</span><span class='op'>)</span> <span class='co'># install.packages("MetBrewer")</span></span>
<span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://here.r-lib.org/'>here</a></span><span class='op'>)</span></span></code></pre>
</div>
</div>
<h2 id="download-files">Download files</h2>
<p>Before we get started, let’s download all of the files you will need for the next three classes.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># conditionally download all of the files used in rmarkdown from github </span></span>
<span><span class='kw'><a href='https://rdrr.io/r/base/source.html'>source</a></span><span class='op'>(</span><span class='st'>"https://raw.githubusercontent.com/rnabioco/bmsc-7810-pbda/main/_posts/2023-12-08-class-8-matricies/download_files.R"</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<h2 id="making-a-heatmap">Making a heatmap</h2>
<p>Today we are going to continue working with the same data we used to talk about matrices and clustering, the top 100 boy and girl names by state for 2020.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>male_names_mat</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.csv</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"class_8-10_data"</span>, <span class='st'>"boy_name_counts.csv"</span><span class='op'>)</span>,</span>
<span>                           row.names <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span></span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>female_names_mat</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.csv</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"class_8-10_data"</span>, <span class='st'>"girl_name_counts.csv"</span><span class='op'>)</span>,</span>
<span>                             row.names <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span></span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>We’ve loaded in both the boy and the girl names as separate matrices, we can combine these using <code>rbind</code> to bind the rows</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>names_mat</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>rbind</a></span><span class='op'>(</span><span class='va'>male_names_mat</span>, <span class='va'>female_names_mat</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>To make a heatmap, we will use the pheatmap package</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>names_mat</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-4-1.png" width="624" /></p>
</div>
<p>Notice here we can only see the names for California</p>
<h3 id="cleaning-up-by-normalizing-values">Cleaning up by normalizing values</h3>
<p>It’s pretty hard to see much structure in the data just using the raw values. Let’s try with our normalized values</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>normalized_mat</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='va'>names_mat</span><span class='op'>)</span> <span class='op'>/</span> <span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>colSums</a></span><span class='op'>(</span><span class='va'>names_mat</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>normalized_mat</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-5-1.png" width="624" /></p>
</div>
<p>You can see that when we try to normalize the values by dividing each number by the total number for the state, the trends in the data are much more clear and we don’t only see the data for California.</p>
<h3 id="cleaning-up-by-scaling-values">Cleaning up by scaling values</h3>
<p>Another popular way to normalize the data for a heatmap is called a z-score. This is also know a mean-centered scaled data. The z-score is calculated as follows</p>
<div>
<span class="math display">\[
zscore = \frac{x - \mu}{\sigma}
\]</span>
<div>

<p>Where x is each value <span class="math inline">\(\mu\)</span> is the mean for the values and <span class="math inline">\(\sigma\)</span> is the standard deviation for the values. Here, <span class="math inline">\(x - \mu\)</span> is called “centering” and dividing by <span class="math inline">\(\sigma\)</span> is called “scaling”</p>
<p>In R, we can scale the data using the <code>scale</code> function. Scaling is done on the columns, so here we can use the original matrix without any transformations</p>
<p><code>scale</code> takes 3 arguments</p>
<pre><code>x - a numeric matrix(like object).

center - either a logical value or numeric-alike vector of length equal to the
number of columns of x, where ‘numeric-alike’ means that as.numeric(.) will be
applied successfully if is.numeric(.) is not true.

scale   - either a logical value or a numeric-alike vector of length equal to the
number of columns of x.</code></pre>
<p>The <code>center</code> refers to <span class="math inline">\(x - \mu\)</span> from the equation above and <code>scale</code> refers to dividing by <span class="math inline">\(\sigma\)</span>. We will set both to be true to be a z-score.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>scaled_mat</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/scale.html'>scale</a></span><span class='op'>(</span><span class='va'>names_mat</span>, scale <span class='op'>=</span> <span class='cn'>TRUE</span>, center <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-7-1.png" width="624" /></p>
</div>
<p>Notice how we again can see more structure in the data.</p>
<p>The scale shows both negative and positive values. Negative values had counts less than the mean, positive values had counts above the mean, and 0 means the value was equal to the mean. An important note here, values that are negative do not mean they were zero. For example, the scaled value for “Grayson” in California is negative:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>scaled_mat</span><span class='op'>[</span><span class='st'>"Grayson"</span>, <span class='st'>"California"</span><span class='op'>]</span></span></code></pre>
</div>
<pre><code>[1] -1.043828</code></pre>
</div>
<p>but there are still several hundred individuals named “Grayson” in California.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>names_mat</span><span class='op'>[</span><span class='st'>"Grayson"</span>, <span class='st'>"California"</span><span class='op'>]</span></span></code></pre>
</div>
<pre><code>[1] 453</code></pre>
</div>
<p><strong>Note if you want a z-score of gene expression data, you will want to center and scale by the genes not the samples. Most gene expression data is stored as gene x sample so you will likely need to transform the matrix before scaling:</strong></p>
<pre><code>scaled_mat &lt;- t(scale(t(unscaled_mat), scale = TRUE, center = TRUE))</code></pre>
<p><strong>Exercise</strong>
Looking at the three heatmaps, what is the most popular name? What heatmap is easiest to interpret?</p>
<h3 id="changing-the-color-palette">Changing the color palette</h3>
<p>The default color palette for pheatmap isn’t always the color palette you want. One color palette I like is the magma color from the <code>viridis</code> package. To use this package you can just call any of the color functions. and provide the number of colors you want in the palette. We will use the <code>magma</code> function below.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>, color <span class='op'>=</span> <span class='fu'><a href='https://sjmgarnier.github.io/viridisLite/reference/viridis.html'>magma</a></span><span class='op'>(</span><span class='fl'>100</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-10-1.png" width="624" /></p>
</div>
<p>You can also add custom color palettes using <code>colorRampPalette</code>. This creates a function to generate a color palette with your colors for any number of values:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>color_function</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/colorRamp.html'>colorRampPalette</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"navy"</span>, <span class='st'>"white"</span>, <span class='st'>"red"</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='va'>color_function</span></span></code></pre>
</div>
<pre><code>function (n) 
{
    x &lt;- ramp(seq.int(0, 1, length.out = n))
    if (ncol(x) == 4L) 
        rgb(x[, 1L], x[, 2L], x[, 3L], x[, 4L], maxColorValue = 255)
    else rgb(x[, 1L], x[, 2L], x[, 3L], maxColorValue = 255)
}
&lt;bytecode: 0x7fdaba45bb70&gt;
&lt;environment: 0x7fdab8051430&gt;</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>color_function</span><span class='op'>(</span><span class='fl'>10</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code> [1] &quot;#000080&quot; &quot;#38389C&quot; &quot;#7171B8&quot; &quot;#AAAAD4&quot; &quot;#E2E2F0&quot; &quot;#FFE2E2&quot;
 [7] &quot;#FFAAAA&quot; &quot;#FF7171&quot; &quot;#FF3838&quot; &quot;#FF0000&quot;</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>, color <span class='op'>=</span> <span class='fu'>color_function</span><span class='op'>(</span><span class='fl'>100</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-13-1.png" width="624" /></p>
</div>
<p>You can also run this in one line of code and not save the function</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>, color <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/colorRamp.html'>colorRampPalette</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"navy"</span>, <span class='st'>"white"</span>, <span class='st'>"red"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>(</span><span class='fl'>100</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-14-1.png" width="624" /></p>
</div>
<p>You can even use custom color palettes with any HEX color codes you want. A personal favorite of mine is a blue/yellow palette from the <code>ArchR</code> package.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>blueYellow</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"#352A86"</span>, <span class='st'>"#343DAE"</span>, <span class='st'>"#0262E0"</span>, <span class='st'>"#1389D2"</span>, <span class='st'>"#2DB7A3"</span>,</span>
<span>                  <span class='st'>"#A5BE6A"</span>, <span class='st'>"#F8BA43"</span>, <span class='st'>"#F6DA23"</span>, <span class='st'>"#F8FA0D"</span><span class='op'>)</span></span>
<span></span>
<span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>, color <span class='op'>=</span> <span class='va'>blueYellow</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-15-1.png" width="624" /></p>
</div>
<p><strong>Exercise</strong>
Make a heatmap with your own color palette</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># TODO make the heatmap with your unique color palette</span></span></code></pre>
</div>
</div>
<h3 id="adding-annotations">Adding annotations</h3>
<p>We can help interpretation by adding row and column annotations to the plot. Let’s first add some row annotations showing if the name is a male or a female name. We start by making a data frame of the names and indicating if the name was from the male or female database.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>names_annotation</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span><span class='st'>"sex"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='st'>"female"</span>, <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>female_names_mat</span><span class='op'>)</span><span class='op'>)</span>,</span>
<span>                                         <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='st'>"male"</span>, <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>male_names_mat</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>,</span>
<span>                               row.names <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>female_names_mat</span><span class='op'>)</span>,</span>
<span>                                             <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>male_names_mat</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span></span>
<span></span>
<span><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>names_annotation</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>             sex
Ava       female
Olivia    female
Emma      female
Charlotte female
Harper    female
Amelia    female</code></pre>
</div>
<p>Note here the row names are the same as the row names of our matrix and we have one column named “sex”</p>
<p>We can now add this to the plot using the <code>annotation_row</code>. The key here is that the row names of your matrix must be the same as the row names of your data frame.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>blueYellow</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"#352A86"</span>, <span class='st'>"#343DAE"</span>, <span class='st'>"#0262E0"</span>, <span class='st'>"#1389D2"</span>, <span class='st'>"#2DB7A3"</span>,</span>
<span>                  <span class='st'>"#A5BE6A"</span>, <span class='st'>"#F8BA43"</span>, <span class='st'>"#F6DA23"</span>, <span class='st'>"#F8FA0D"</span><span class='op'>)</span></span>
<span></span>
<span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>,</span>
<span>         color <span class='op'>=</span> <span class='va'>blueYellow</span>,</span>
<span>         annotation_row <span class='op'>=</span> <span class='va'>names_annotation</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-18-1.png" width="624" /></p>
</div>
<p>Now you can see very clearly what names were in the female database and what names were in the male database. Again, we can change the default colors by using a <code>list</code> object. Here the names of the <code>list</code> must match the column names in your data frame and the names of the colors in the list must match the levels of that column.</p>
<p>We can either use the name of colors:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>all_colors</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='st'>"sex"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"male"</span> <span class='op'>=</span> <span class='st'>"red"</span>, <span class='st'>"female"</span> <span class='op'>=</span> <span class='st'>"yellow"</span><span class='op'>)</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>all_colors</span></span></code></pre>
</div>
<pre><code>$sex
    male   female 
   &quot;red&quot; &quot;yellow&quot; </code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>,</span>
<span>         color <span class='op'>=</span> <span class='va'>blueYellow</span>,</span>
<span>         annotation_row <span class='op'>=</span> <span class='va'>names_annotation</span>,</span>
<span>         annotation_colors <span class='op'>=</span> <span class='va'>all_colors</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-20-1.png" width="624" /></p>
</div>
<p>Or a HEX code:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>all_colors</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='st'>"sex"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"male"</span> <span class='op'>=</span> <span class='st'>"#B067A3"</span>, <span class='st'>"female"</span> <span class='op'>=</span> <span class='st'>"#9C954D"</span><span class='op'>)</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>all_colors</span></span></code></pre>
</div>
<pre><code>$sex
     male    female 
&quot;#B067A3&quot; &quot;#9C954D&quot; </code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>,</span>
<span>         color <span class='op'>=</span> <span class='va'>blueYellow</span>,</span>
<span>         annotation_row <span class='op'>=</span> <span class='va'>names_annotation</span>,</span>
<span>         annotation_colors <span class='op'>=</span> <span class='va'>all_colors</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-22-1.png" width="624" /></p>
</div>
<p><strong>Exercise</strong>
Add your own colors to the row annotations</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># TODO Add your own row annotaiton colors</span></span></code></pre>
</div>
</div>
<p>We can also add annotations to the columns using the same approach. Let’s load in some of the data I’ve compiled for the states</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>state_info</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.csv</a></span><span class='op'>(</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"class_8-10_data"</span>, <span class='st'>"state_info.csv"</span><span class='op'>)</span><span class='op'>)</span></span>
<span></span>
<span><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>state_info</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>       State Location   Size Density
1    Alabama    South Middle       4
2     Alaska     West  Small       6
3    Arizona     West Middle       4
4   Arkansas    South Middle       5
5 California     West  Large       2
6   Colorado     West Middle       5</code></pre>
</div>
<p>This table has information about the location, the size (based on population as small middle or large) and the density (on a scale of 1-6).</p>
<p>We first will need to reformat so that the row names are the sates. Now the row names must be the same as the column names in the matrix.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>state_info</span> <span class='op'>&lt;-</span> <span class='fu'>tibble</span><span class='fu'>::</span><span class='fu'><a href='https://tibble.tidyverse.org/reference/rownames.html'>column_to_rownames</a></span><span class='op'>(</span><span class='va'>state_info</span>, <span class='st'>"State"</span><span class='op'>)</span></span>
<span></span>
<span></span>
<span><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>state_info</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>           Location   Size Density
Alabama       South Middle       4
Alaska         West  Small       6
Arizona        West Middle       4
Arkansas      South Middle       5
California     West  Large       2
Colorado       West Middle       5</code></pre>
</div>
<p>We can now pass this data frame to <code>annotation_col</code> in <code>pheatmap</code></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>,</span>
<span>         color <span class='op'>=</span> <span class='va'>blueYellow</span>,</span>
<span>         annotation_col <span class='op'>=</span> <span class='va'>state_info</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-26-1.png" width="624" /></p>
</div>
<p>You can see that we now have three levels of annotation on our heatmap. One thing to notice is that our population density is given as a scale of colors because it is seen as a numeric score. If we instead use this as a factor, this scale will go away:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>state_info</span><span class='op'>$</span><span class='va'>Density</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>state_info</span><span class='op'>$</span><span class='va'>Density</span><span class='op'>)</span></span>
<span></span>
<span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>,</span>
<span>         color <span class='op'>=</span> <span class='va'>blueYellow</span>,</span>
<span>         annotation_col <span class='op'>=</span> <span class='va'>state_info</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-27-1.png" width="624" /></p>
</div>
<p>As before, we can add in our own color palettes by using a named list where the list names correspond to the column names of our data frame. Here, I’m going to use the package <code>MetBrewer</code> to generate palettes for me. <a href="https://github.com/BlakeRMills/MetBrewer">You can see all possible palette options here</a></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>location_colors</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/MetBrewer/man/met.brewer.html'>met.brewer</a></span><span class='op'>(</span><span class='st'>"Pissaro"</span>,</span>
<span>                              n <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>state_info</span><span class='op'>$</span><span class='va'>Location</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span></span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/character.html'>as.character</a></span><span class='op'>(</span><span class='op'>)</span></span>
<span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>location_colors</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>state_info</span><span class='op'>$</span><span class='va'>Location</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>location_colors</span></span></code></pre>
</div>
<pre><code>      South        West New_England     Midwest 
  &quot;#4c825d&quot;   &quot;#8cae9e&quot;   &quot;#8dc7dc&quot;   &quot;#0e2a4d&quot; </code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>size_colors</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/MetBrewer/man/met.brewer.html'>met.brewer</a></span><span class='op'>(</span><span class='st'>"Navajo"</span>,</span>
<span>                          n <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>state_info</span><span class='op'>$</span><span class='va'>Size</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span></span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/character.html'>as.character</a></span><span class='op'>(</span><span class='op'>)</span></span>
<span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>size_colors</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>state_info</span><span class='op'>$</span><span class='va'>Size</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>size_colors</span></span></code></pre>
</div>
<pre><code>   Middle     Small     Large 
&quot;#660d20&quot; &quot;#e59a52&quot; &quot;#edce79&quot; </code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>density_colors</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/MetBrewer/man/met.brewer.html'>met.brewer</a></span><span class='op'>(</span><span class='st'>"Juarez"</span>,</span>
<span>                             n <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>state_info</span><span class='op'>$</span><span class='va'>Density</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span></span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/character.html'>as.character</a></span><span class='op'>(</span><span class='op'>)</span></span>
<span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>density_colors</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>state_info</span><span class='op'>$</span><span class='va'>Density</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>density_colors</span></span></code></pre>
</div>
<pre><code>        1         2         3         4         5         6 
&quot;#a82203&quot; &quot;#208cc0&quot; &quot;#f1af3a&quot; &quot;#cf5e4e&quot; &quot;#637b31&quot; &quot;#003967&quot; </code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>all_colors</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='st'>"Location"</span> <span class='op'>=</span> <span class='va'>location_colors</span>,</span>
<span>                   <span class='st'>"Size"</span> <span class='op'>=</span> <span class='va'>size_colors</span>,</span>
<span>                   <span class='st'>"Density"</span> <span class='op'>=</span> <span class='va'>density_colors</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>all_colors</span></span></code></pre>
</div>
<pre><code>$Location
      South        West New_England     Midwest 
  &quot;#4c825d&quot;   &quot;#8cae9e&quot;   &quot;#8dc7dc&quot;   &quot;#0e2a4d&quot; 

$Size
   Middle     Small     Large 
&quot;#660d20&quot; &quot;#e59a52&quot; &quot;#edce79&quot; 

$Density
        1         2         3         4         5         6 
&quot;#a82203&quot; &quot;#208cc0&quot; &quot;#f1af3a&quot; &quot;#cf5e4e&quot; &quot;#637b31&quot; &quot;#003967&quot; </code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>,</span>
<span>         color <span class='op'>=</span> <span class='va'>blueYellow</span>,</span>
<span>         annotation_col <span class='op'>=</span> <span class='va'>state_info</span>,</span>
<span>         annotation_colors <span class='op'>=</span> <span class='va'>all_colors</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-29-1.png" width="624" /></p>
</div>
<p>Finally, we can add in the annotation from the rows as well. First we can add in the colors we had before</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>all_colors</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>all_colors</span>, </span>
<span>                <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='st'>"sex"</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"male"</span> <span class='op'>=</span> <span class='st'>"#B067A3"</span>, <span class='st'>"female"</span> <span class='op'>=</span> <span class='st'>"#9C954D"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>all_colors</span></span></code></pre>
</div>
<pre><code>$Location
      South        West New_England     Midwest 
  &quot;#4c825d&quot;   &quot;#8cae9e&quot;   &quot;#8dc7dc&quot;   &quot;#0e2a4d&quot; 

$Size
   Middle     Small     Large 
&quot;#660d20&quot; &quot;#e59a52&quot; &quot;#edce79&quot; 

$Density
        1         2         3         4         5         6 
&quot;#a82203&quot; &quot;#208cc0&quot; &quot;#f1af3a&quot; &quot;#cf5e4e&quot; &quot;#637b31&quot; &quot;#003967&quot; 

$sex
     male    female 
&quot;#B067A3&quot; &quot;#9C954D&quot; </code></pre>
</div>
<p>And then add the <code>names_annotation</code> back into the argument for <code>annotation_row</code></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>,</span>
<span>         color <span class='op'>=</span> <span class='va'>blueYellow</span>,</span>
<span>         annotation_col <span class='op'>=</span> <span class='va'>state_info</span>,</span>
<span>         annotation_row <span class='op'>=</span> <span class='va'>names_annotation</span>,</span>
<span>         annotation_colors <span class='op'>=</span> <span class='va'>all_colors</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-31-1.png" width="624" data-distill-preview=1 /></p>
</div>
<p><strong>Exercise</strong>
What happens if <code>all_colors</code> only has some of the colors in the annotation data frames?</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># TODO try making the heatmap with a list that only contains color arguments for</span></span>
<span><span class='co'># some of the columns in the annotation data frames</span></span></code></pre>
</div>
</div>
<p>Notice how the <code>annotation_colors</code> argument accepts the color arguments for both the rows and the columns. The important things about this argument to color the annotation is</p>
<ol type="1">
<li>It must be a list</li>
<li>The names of the list must be in the columns for the <code>annotation_row</code> or <code>annotation_col</code> data frame</li>
<li>The names of the colors must match the values in the matching column</li>
<li>You do not need to have all columns present in your color list - any missing columns will be given the default colors</li>
</ol>
<p>There are also a few important aspects of the annotation data frames
1. The rownames of <code>annotation_row</code> must match the rownames of the matrix (although the order does not need to be the same)
2. The rownames of <code>annotation_col</code> must match the column names of the matrix (although the order does not need to be the same)
3. You can add as many annotations to the rows and columns that you want. You just need to include these as columns in either the <code>annotation_row</code> or <code>annotation_col</code> data frame.</p>
<h3 id="clustering">Clustering</h3>
<p>One aspect of the plots that you’ve probably noticed is dendrograms for both the rows and the columns. The clustering is done using <code>hclust</code>like we did in the clustering lecture.</p>
<p>Just like with our clustering methods we can cut the dendrogram based on an expected number of clusters to group either the states or the names. <code>pheatmap</code> has a function that uses <code>cutree</code> to identify clusters and physically separates these clusters with a white space using <code>cutree_rows</code> and <code>cutree_cols</code>. First, we can visualize three clusters of the names.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>,</span>
<span>         color <span class='op'>=</span> <span class='va'>blueYellow</span>,</span>
<span>         annotation_col <span class='op'>=</span> <span class='va'>state_info</span>,</span>
<span>         annotation_row <span class='op'>=</span> <span class='va'>names_annotation</span>,</span>
<span>         annotation_colors <span class='op'>=</span> <span class='va'>all_colors</span>,</span>
<span>         cutree_rows <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-33-1.png" width="624" /></p>
</div>
<p>Or we can visualize three clusters of the states</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>,</span>
<span>         color <span class='op'>=</span> <span class='va'>blueYellow</span>,</span>
<span>         annotation_col <span class='op'>=</span> <span class='va'>state_info</span>,</span>
<span>         annotation_row <span class='op'>=</span> <span class='va'>names_annotation</span>,</span>
<span>         annotation_colors <span class='op'>=</span> <span class='va'>all_colors</span>,</span>
<span>         cutree_cols <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-34-1.png" width="624" /></p>
</div>
<p>We could even visuzalize both at once</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>,</span>
<span>         color <span class='op'>=</span> <span class='va'>blueYellow</span>,</span>
<span>         annotation_col <span class='op'>=</span> <span class='va'>state_info</span>,</span>
<span>         annotation_row <span class='op'>=</span> <span class='va'>names_annotation</span>,</span>
<span>         annotation_colors <span class='op'>=</span> <span class='va'>all_colors</span>,</span>
<span>         cutree_rows <span class='op'>=</span> <span class='fl'>3</span>,</span>
<span>         cutree_cols <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-35-1.png" width="624" /></p>
</div>
<p><strong>Exercise</strong>
Try making the heatmap with different numbers of clusters. What number of clusters makes the most sense?</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># TODO Make the heatmap cutting to make different numbers of clusters</span></span></code></pre>
</div>
</div>
<p>As with using <code>cutree</code> ourselves, you can pick any number of clusters to visualize in this way.</p>
<p>In addition to visualizing these clusters, we can also pull the clustering information out of the heatmap object. First we can save the plot to a variable. Right now we are setting <code>silent</code> to <code>TRUE</code> so the heatmap isn’t drawn</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>heatmap</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>,</span>
<span>                    color <span class='op'>=</span> <span class='va'>blueYellow</span>,</span>
<span>                    annotation_col <span class='op'>=</span> <span class='va'>state_info</span>,</span>
<span>                    annotation_row <span class='op'>=</span> <span class='va'>names_annotation</span>,</span>
<span>                    annotation_colors <span class='op'>=</span> <span class='va'>all_colors</span>,</span>
<span>                    cutree_rows <span class='op'>=</span> <span class='fl'>3</span>,</span>
<span>                    cutree_cols <span class='op'>=</span> <span class='fl'>3</span>,</span>
<span>                    silent <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></span>
<span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>heatmap</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>[1] &quot;tree_row&quot; &quot;tree_col&quot; &quot;kmeans&quot;   &quot;gtable&quot;  </code></pre>
</div>
<p>The <code>hclust</code> object for the row are in <code>tree_row</code> and the <code>hclust</code> object for the column are in <code>tree_col</code></p>
<p>We can plot just the dendrogram as we did previously</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>heatmap</span><span class='op'>$</span><span class='va'>tree_col</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-38-1.png" width="624" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>heatmap</span><span class='op'>$</span><span class='va'>tree_row</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-39-1.png" width="1152" /></p>
</div>
<p>We can also now use <code>cutree</code> on these <code>hclust</code> objects in the exact same way we did in the clustering lecture to pull out clusters</p>
<p>Columns:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/stats/cutree.html'>cutree</a></span><span class='op'>(</span>tree <span class='op'>=</span> <span class='va'>heatmap</span><span class='op'>$</span><span class='va'>tree_col</span>, k <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>    Alabama      Alaska     Arizona    Arkansas  California 
          1           2           3           1           3 
   Colorado Connecticut    Delaware     Florida     Georgia 
          2           3           3           3           1 
     Hawaii       Idaho    Illinois     Indiana        Iowa 
          3           2           3           2           2 
     Kansas    Kentucky   Louisiana       Maine    Maryland 
          2           1           1           2           3 </code></pre>
</div>
<p>Rows:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/stats/cutree.html'>cutree</a></span><span class='op'>(</span>tree <span class='op'>=</span> <span class='va'>heatmap</span><span class='op'>$</span><span class='va'>tree_row</span>, k <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>  William     James      John    Elijah      Noah      Liam     Mason 
        1         1         2         1         1         1         2 
   Oliver     Henry   Jackson    Samuel     Jaxon     Asher   Grayson 
        1         2         2         3         3         2         2 
     Levi   Michael    Carter  Benjamin   Charles     Wyatt    Thomas 
        2         2         2         2         3         2         3 
    Aiden      Luke     David      Owen    Daniel     Logan    Joseph 
        3         3         3         2         2         2         3 
    Lucas    Joshua      Jack Alexander  Maverick   Gabriel     Ethan 
        2         3         2         2         3         3         2 
      Eli     Isaac    Hunter      Ezra  Theodore       Ava    Olivia 
        3         3         3         3         2         1         1 
     Emma Charlotte    Harper    Amelia Elizabeth    Evelyn  Isabella 
        1         1         2         1         3         2         2 
     Ella     Avery   Abigail    Sophia     Layla       Mia   Madison 
        3         3         3         2         3         2         3 
     Lily     Ellie      Nova   Eleanor      Zoey  Brooklyn     Riley 
        3         3         3         3         3         3         3 
     Nora      Aria      Mila    Stella   Natalie      Luna  Penelope 
        3         3         3         3         3         3         3 
   Aurora    Claire 
        3         3 </code></pre>
</div>
<p><strong>Exercise</strong>
What names are the most similar? What states are the most similar?</p>
<p>Another option is to remove the clustering all together using <code>cluster_rows</code> and <code>cluster_cols</code></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>,</span>
<span>         color <span class='op'>=</span> <span class='va'>blueYellow</span>,</span>
<span>         annotation_col <span class='op'>=</span> <span class='va'>state_info</span>,</span>
<span>         annotation_row <span class='op'>=</span> <span class='va'>names_annotation</span>,</span>
<span>         annotation_colors <span class='op'>=</span> <span class='va'>all_colors</span>,</span>
<span>         cluster_rows <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span>         cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-42-1.png" width="624" /></p>
</div>
<p>If we don’t cluster, the order of the rows and columns is taken directly from the order of the matrix. We can change to alphabetical order as follows</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>scaled_mat_ord</span> <span class='op'>&lt;-</span> <span class='va'>scaled_mat</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>scaled_mat</span><span class='op'>)</span><span class='op'>)</span>, <span class='op'>]</span></span>
<span></span>
<span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat_ord</span>,</span>
<span>         color <span class='op'>=</span> <span class='va'>blueYellow</span>,</span>
<span>         annotation_col <span class='op'>=</span> <span class='va'>state_info</span>,</span>
<span>         annotation_row <span class='op'>=</span> <span class='va'>names_annotation</span>,</span>
<span>         annotation_colors <span class='op'>=</span> <span class='va'>all_colors</span>,</span>
<span>         cluster_rows <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span>         cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-43-1.png" width="624" /></p>
</div>
<p>We could even order the states by the density</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>state_info</span> <span class='op'>&lt;-</span> <span class='va'>state_info</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span></span>
<span>  <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/arrange.html'>arrange</a></span><span class='op'>(</span><span class='va'>Density</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>scaled_mat_ord</span> <span class='op'>&lt;-</span> <span class='va'>scaled_mat</span><span class='op'>[</span> , <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/match.html'>match</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>scaled_mat</span><span class='op'>)</span>,</span>
<span>                                            <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>state_info</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span><span class='op'>]</span></span>
<span></span>
<span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat_ord</span>,</span>
<span>         color <span class='op'>=</span> <span class='va'>blueYellow</span>,</span>
<span>         annotation_col <span class='op'>=</span> <span class='va'>state_info</span>,</span>
<span>         annotation_row <span class='op'>=</span> <span class='va'>names_annotation</span>,</span>
<span>         annotation_colors <span class='op'>=</span> <span class='va'>all_colors</span>,</span>
<span>         cluster_rows <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span>         cluster_cols <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-44-1.png" width="624" /></p>
</div>
<p><strong>Exercise</strong>
Order the heatmap by density, but cluster the names</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># TODO order the heatmap by density but cluster the names</span></span></code></pre>
</div>
</div>
<p><strong>Exercise</strong>
Order the heatmap by location, but cluster the names</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># TODO order the heatmap by location and cluster the names</span></span></code></pre>
</div>
</div>
<h3 id="remove-column-and-row-names">Remove column and row names</h3>
<p>One other helpful way to adjust your heatmap is to remove the row and column names. You can control if the row and column names are seen using <code>show_rownames</code> and <code>show_colnames</code></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/pheatmap/man/pheatmap.html'>pheatmap</a></span><span class='op'>(</span><span class='va'>scaled_mat</span>,</span>
<span>         color <span class='op'>=</span> <span class='va'>blueYellow</span>,</span>
<span>         annotation_col <span class='op'>=</span> <span class='va'>state_info</span>,</span>
<span>         annotation_row <span class='op'>=</span> <span class='va'>names_annotation</span>,</span>
<span>         annotation_colors <span class='op'>=</span> <span class='va'>all_colors</span>,</span>
<span>         show_rownames <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span>         show_colnames <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="class-10-heatmap_files/figure-html5/unnamed-chunk-47-1.png" width="624" /></p>
</div>
<h3 id="other-aesthetics">Other aesthetics</h3>
<p>There are many possible adjustments you can make, we’ve just gone through the adjustments I use the most often. To see all possible arguments see <code>?pheatmap</code></p>
<h2 id="acknowldgements-and-additional-references">Acknowldgements and additional references</h2>
<p>The content of this class borrows heavily from previous tutorials:</p>
<p><a href="https://davetang.org/muse/2018/05/15/making-a-heatmap-in-r-with-the-pheatmap-package/">making heatmaps</a></p>
<h2 id="im-a-biologist-why-should-i-care">I’m a biologist, why should I care?</h2>
<ul>
<li>Heatmaps are common tools to visualize data
<ul>
<li>They are frequently used for sequencing data - RNA-seq, scRNA-seq, ATAC-seq</li>
<li>They are a good way of showing lots of data</li>
<li>Because of the scaling, they can be misleading so it’s good to know how to interpret them</li>
</ul></li>
<li>You will likely encounter heatmaps in papers that you read. My goal is to improve your understanding of the techniques so you can better interpret them on your own</li>
</ul>
<div class="sourceCode" id="cb20"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom"></div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
