---
title: "Class 3: Reshaping data into a tidy format"
author:
  - name: "Kent Riemondy"
    url: https://github.com/kriemo
    affiliation: "RNA Bioscience Initiative"
    affiliation_url: https://medschool.cuanschutz.edu/rbi
    orcid_id: 0000-0003-0750-1273
output:
  distill::distill_article:
    self_contained: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE, 
                      warning = FALSE,
                      R.options = list(width = 80))
```

```{r, echo = FALSE}
# conditionally download all of the files used in rmarkdown from github 
library(here)
data_dir <- here("_posts/2022-11-15-class-3-reshaping-data-into-a-tidy-format/data")
dir.create(data_dir, showWarnings = FALSE)

files_to_dl <- list(
  "income_per_person.csv" = "https://github.com/rnabioco/bmsc-7810-pbda/raw/main/data/class3/income_per_person.csv",
  "dmel_tx_info.csv.gz" = "https://github.com/rnabioco/bmsc-7810-pbda/raw/main/data/class3/dmel_tx_info.csv.gz",
  "dmel_peptides_lifecycle.csv.gz" = "https://github.com/rnabioco/bmsc-7810-pbda/raw/main/data/class3/dmel_peptides_lifecycle.csv.gz"
)

missing_files <- files_to_dl[!file.exists(file.path(data_dir, names(files_to_dl)))]

tout_setting <- getOption("timeout")
options(timeout = 6000)

for(id in names(missing_files)){
  download.file(missing_files[[id]], file.path(data_dir, id))
}

options(timeout = tout_setting)
```

*The Rmarkdown for this class is [on github]( https://github.com/rnabioco/bmsc-7810-pbda/blob/main/_posts/2022-11-15-class-3-reshaping-data-into-a-tidy-format/class3.Rmd)*


## Goals for today

- Discuss wide and long (tidy) data representations for analysis 
- Introduce tidyr for "tidying" rectangular data
- Joining related tables with dplyr
- Strategies for missing data


>"Data Scientists spend up to 80% of the time on data cleaning and 20 percent of their time on actual data analysis."
-- Exploratory Data Mining and Data Cleaning. Dasu and Johnson


## Wide versus long data formats

Data can be represented in multiple formats. Today we will discuss two common tabular formats for organizing data for analysis in R. Consider the following dataset, which contains GDP estimates per person for countries throughout history. This representation of the data is commonly referred to as 'wide' data format, which is a matrix-like format containing samples or features as rows or columns, with values associated with each sample and feature. 

```{r, echo = FALSE}
library(here)
library(readr)
gdp_data <- read_csv("data/income_per_person.csv")
gdp_data
```

The wide matrix-like format is very useful and a common format used for statistics and machine learning. Matrices can take advantage of optimized numerical routines and are the data representation of mathematical matrices. We will work with matrices latter in class, particularly their use to generate heatmaps. 

Representing data in a matrix has a few practical implications:  

1) There is only 1 type of data stored in matrix (e.g. each cell is the same unit of observation, the GDP per person). To store more values you need multiple matrices.  

2) The format is not easily manipulated with dplyr/tidyverse.

We will instead start by with working with data in a long (also known as tidy) format. 

The same data could be instead formatted in a long (aka "tidy") format.

```{r, echo = FALSE}
tidyr::pivot_longer(gdp_data, cols = -country, names_to = "year", values_to = "gdp")
```

The long format of this data simplifies the many columns of a matrix into a 3 column data.frame containing 3 variables (`country`, `year`, and `gdp`).

## Tidy data format

>“Tidy datasets are all alike, but every messy dataset is messy in its own way.” –– Hadley Wickham

Tidy data is a concept introduced by Hadley Wickham, chief scientist at Rstudio, and developer of many of the tidyverse packages. A tidy dataset is structured in a manner to be most effectively processed in R using the tidyverse. It makes the data easier to plot, easier to perform computations and perform complex tasks.  

Most data tables that you've worked with are probably not tidy. It takes experience to understand the best way to format the data. As you work more in R and the tidyverse this will become more natural. 

Tidy data has the following attributes:

1) Each variable must have its own column.  
2) Each observation must have its own row.  
3) Each value must have its own cell.  

What is a variable, what is an observation, and what is a value?

1) A value is a number or word.  

2) Every value belongs to a variable and an observation  

3) A variable contains all values that measure the same attribute (e.g. height, temperature, duration, magnitude) across units.  

4) An observation contains all values measured on the same unit (e.g. biological sample, day, gene, country).  


```{r, echo = FALSE}
knitr::include_graphics("https://raw.githubusercontent.com/rnabioco/bmsc-7810-pbda/main/_posts/2022-11-15-class-3-reshaping-data-into-a-tidy-format/img/tidydata.png")
```

Shown below is a simple data table in a tidy format, provided by the tidyr package. 

```{r}
library(tidyr)
table1
```

The same data, represented in wide, matrix-like format, would require 2 tables:

e.g a matrix with the `cases` values per country.

```{r}
table4a
```
e.g a matrix with the `population` values per country   

```{r}
table4b
```

What advantages does the tidy format provide? 

1) Easy to generate summaries of the data. 
   e.g. via group_by() -> summarize()  
   
2) Easy to plot the data using the ggplot2 framework (more on that in later classes)

3) Very easy to join multiple data tables based on key values.

Some disadvantages: 

1) Not space efficient   
2) Not intuitive   
3) Doesn't work with typical machine learning and statistical approaches.  


## Converting between long and wide formats using tidyr 

The [`tidyr`](https://tidyr.tidyverse.org/) package provides functionality to convert datasets into tidy formats. 

- pivot_longer(): convert wide data to long data  
- pivot_wider(): convert long data to wide data  
- separate(): split a single column into multiple columns  
- unite(): combine multiple columns into single columns  

## Reshaping wide data to long

The `pivot_longer` function requires specifying the columns to pivot using the [`tidyselect`](https://tidyr.tidyverse.org/reference/tidyr_tidy_select.html) syntax. This syntax is used elsewhere in the tidyverse and is a useful shorthand to avoid listing all columns of interest.   

`pivot_longer(tbl, cols = <...>)`

```{r, echo = FALSE, fig.cap= "Tables from tidyr cheatsheet from https://posit.co/wp-content/uploads/2022/10/tidyr.pdf" }
knitr::include_graphics("https://github.com/rnabioco/bmsc-7810-pbda/raw/main/_posts/2022-11-15-class-3-reshaping-data-into-a-tidy-format/img/pivot_longer.png")
```


```{r}
#pivot_longer(table4a, cols = `1999`:`2000`) # pivot columns from 1999 -> 2000
pivot_longer(table4a, cols = -country) # pivot all columns not matching country
```
## Reshaping long data to wide

`pivot_wider(tbl, names_from = <...>, values_from = <...>)`

`names_from`: the column whose values will become new columns in the result.    
`values_from`: the column whose values will be in the new columns.  

```{r, echo = FALSE }
knitr::include_graphics("https://github.com/rnabioco/bmsc-7810-pbda/raw/main/_posts/2022-11-15-class-3-reshaping-data-into-a-tidy-format/img/pivot_wider.png")
```

```{r}
pivot_wider(table2, names_from = type, values_from = count)
```


## Separate

`separate` is useful for dealing with data in which a single column contains multiple variables.

`seperate(tbl, col = <...>, into = c(<..., ..., ...>), sep = "...")`

`col`: column to split into multiple columns  
`into`: column names of new columns to be generated, supplied as a character vector (use quotes).  
`sep`: the separator used to split values in the `col` column. Can be a character (`_`) or a integer to indicate the character position to split (2).  

```{r, echo = FALSE}
knitr::include_graphics("https://github.com/rnabioco/bmsc-7810-pbda/raw/main/_posts/2022-11-15-class-3-reshaping-data-into-a-tidy-format/img/separate.png")
```

```{r}
separate(table3, col = rate, into = c("cases", "pop"))
```

## unite 

`unite` is the inverse operation of `separate` and will join multiple columns together. You can also use `str_c` from stringr or `paste()` from base R with mutate.

```{r, echo = FALSE}
knitr::include_graphics("https://github.com/rnabioco/bmsc-7810-pbda/raw/main/_posts/2022-11-15-class-3-reshaping-data-into-a-tidy-format/img/unite.png")
```

```{r}
unite(table5, "year", century:year, sep = "")
#mutate(table5, year = str_c(century, year))
```

## Exercises

1) Tidy the GDP data. 
2) Which country had the highest GDP per person in 1985?
3) What was the mean worldwide GDP in the year 1999?
4) Which country had the highest average GDP per person in the 19th century?

```{r}
library(readr)
read_csv("data/income_per_person.csv")
```


## Binds/Joins

### column binds


```{r, echo = FALSE, fig.cap = "from the dplyr cheatsheet at https://posit.co/wp-content/uploads/2022/10/data-transformation-1.pdf"}
knitr::include_graphics("https://github.com/rnabioco/bmsc-7810-pbda/raw/main/_posts/2022-11-15-class-3-reshaping-data-into-a-tidy-format/img/bind_cols.png")
```

`bind_cols(tbl_1, tbl_2, ...)`

`bind_cols` will bind the columns from 2 or more tables into 1 table. Note that with binds you need to ensure that each table has the same number of rows.

```{r}
library(dplyr)
bind_cols(tibble(x = 1:3), tibble(y = 3:5))
```

### row binds

`bind_rows` binds rows from multiple tables into one table. Unlike with `bind_cols` the number of columns doesn't need to match when using `bind_rows`. 

`bind_rows(tbl_1, tbl_2, ..., .id = NULL)`

```{r, echo = FALSE}
knitr::include_graphics("https://github.com/rnabioco/bmsc-7810-pbda/raw/main/_posts/2022-11-15-class-3-reshaping-data-into-a-tidy-format/img/bind_rows.png")
```

```{r}
one <- starwars[1:4, 1:4]
two <- starwars[9:12, 1:4]

bind_rows(one, two)
```

You can also use a list of `data.frames` with `bind_rows`. If the list is named, you can use the `.id` argument to store a column specifying the name of the data.frame in the output. 

```{r}
lst_of_dfs <- list(a = one,
                   b = two)
bind_rows(lst_of_dfs)

bind_rows(lst_of_dfs, .id = "source_table")
```

## Joins

Join operations are used to join one table with another table by matching the values shared in particular columns. They are used join datasets with a shared variable. 

There are multiple way to join two tables, depending on how you want to handle different combinations of values present in two tables.

From the dplyr documentation:

```
The mutating joins add columns from y to x, matching rows based on the keys:

inner_join(): includes all rows in x and y.

left_join(): includes all rows in x.

right_join(): includes all rows in y.

full_join(): includes all rows in x or y.

If a row in x matches multiple rows in y, all the rows in y will 
be returned once for each matching row in x.
```

Consider the following simple tables:

```{r}
band_members
```

```{r}
band_instruments
```


```{r}
inner_join(band_members, band_instruments)
left_join(band_members, band_instruments)
full_join(band_members, band_instruments)
```

The `Joining, by = "name"` message indicates which columns were used to determine matching rows between the two tables. This is auto-detected based on shared column names. You can use the `by` argument to explicitly specify the columns you'd like to join, which is useful if the columns of interest have different names in the two tables. 

## Missing data

### Zeroes, `NA`, `NaN` and `NULL`

+ Don't use use zeroes to represent missing data. `0` is valid observed value.

+ `NA` (Not Available) is most often use to represent missing data.

+ `NaN` (Not a Number) is the result of an undefined operation, e.g. `0 / 0`.

+ `NULL` means "undefined" and is only used in a programming context (i.e., a function that returns `NULL`). You can't put `NULL` values in a data frame.

Let's examine a data frame with some missing data.

```{r complete}
starwars
```

### `filter` with `is.na()`

You can identify variables with `NA` values by combining `filter()` and `is.na()`.

```{r filter_na, eval = FALSE}
# find rows where mass is NA
filter(starwars, is.na(mass))

# find rows where mass is *not* NA
filter(starwars, !is.na(mass))
```

### `na.omit()`

You can remove **all** rows containing `NA` values with `na.omit()`.

```{r na.omit, eval = FALSE}
na.omit(starwars)
```

### Computing with `NA` values

Exclude `NA` values from operations with `na.rm = TRUE`.

```{r stat_na, eval = FALSE}
starwars$mass
# if NAs are present, the result is NA
sum(starwars$mass)
# solution: drop NAs from the calculation
sum(starwars$mass, na.rm = TRUE)
```

```{r, eval = FALSE}
group_by(starwars, species) %>% 
  summarize(avg_mass = mean(mass, na.rm = TRUE))
```

## Exercise

For this exercise we will tidy a dataset derived from this publication:

>Casas-Vila N, Bluhm A, Sayols S, Dinges N, Dejung M, Altenhein T, Kappei D, Altenhein B, Roignant JY, Butter F. The developmental proteome of Drosophila melanogaster. Genome Res. 2017 Jul;27(7):1273-1285. doi: 10.1101/gr.213694.116. Epub 2017 Apr 5. PMID: 28381612; PMCID: PMC5495078.
>

This is a proteomics dataset generated from Drosophila melanogaster. Samples were collected through the fly lifespan, with 4 samples taken per timepoint.

```{r, echo = FALSE}
knitr::include_graphics("https://github.com/rnabioco/bmsc-7810-pbda/raw/main/_posts/2022-11-15-class-3-reshaping-data-into-a-tidy-format/img/fly_proteome.png")
```

The dataset provided here contains > 100 columns and > 50k rows. The first column (`Sequence`) is the amino acid sequence of the peptide measured, the second column is the transcript id of the protein that the peptide matches (`Proteins`), and the remaining columns contain intensity values indicating the amount of each peptide in each sample (arbitrary units). 

The sample names are coded in each column as follows:
`timepoint_replicate`
`e02` - `e20`: embryonic stages
`L1` - `L3l`: larval stages
`p1` - `p5`: pupal stages
`vm` and `vf`: virgin male and virgin female flies ("adult young")
`m` and `f`: adult flies

1) Tidy this data
2) Average the intensity values per protein at each time point 
3) What is the most abundant protein in the adult female fly?

```r
prot_dat <- read_csv("data/dmel_peptides_lifecycle.csv.gz")
...
...
```

4) Next we are going to add some additional information to make these data more interpretable. The `dmel_tx_info.csv.gz` file contains information about each transcript. Use a `join` operation to join this information with the fly protein abundance dataset. Note that the `Proteins` column matches the information in the `flybase_transcript_id` column.  

```r
dmel_tx_info <- read_csv("data/dmel_tx_info.csv.gz")
...
...

```

5) What is the name and description of the most abundant protein found in the earliest embryonic stage (`e02`)?



<details><summary>Show session info</summary>

```{r code}
sessionInfo()
```
</details>

## Acknowledgements and additional references {.appendix}

The content of this class borrows heavily from previous tutorials:

Tutorial organization:
https://github.com/sjaganna/molb7910-2019 

R tutorials and documentation:
https://github.com/tidyverse/dplyr 
https://r4ds.had.co.nz/index.html 
