<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>


  <!--radix_placeholder_meta_tags-->
  <title>Class wrap up: Data analysis, tips and resources</title>



  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2022-12-05"/>
  <meta property="article:created" itemprop="dateCreated" content="2022-12-05"/>
  <meta name="article:author" content="Rui Fu, Kent Riemondy"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="Class wrap up: Data analysis, tips and resources"/>
  <meta property="og:type" content="article"/>
  <meta property="og:locale" content="en_US"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="Class wrap up: Data analysis, tips and resources"/>

  <!--/radix_placeholder_meta_tags-->
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","author","date","output"]}},"value":[{"type":"character","attributes":{},"value":["Class wrap up: Data analysis, tips and resources"]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name"]}},"value":[{"type":"character","attributes":{},"value":["Rui Fu, Kent Riemondy"]}]}]},{"type":"character","attributes":{},"value":["2022-12-05"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc"]}},"value":[{"type":"logical","attributes":{},"value":[false]},{"type":"logical","attributes":{},"value":[true]}]}]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["img/git-basics.png","img/github-flow.png","img/traceback.png","resources_files/anchor-4.2.2/anchor.min.js","resources_files/bowser-1.9.3/bowser.min.js","resources_files/d3-3.5.6/d3.min.js","resources_files/d3-3.5.6/LICENSE","resources_files/distill-2.2.21/template.v2.js","resources_files/header-attrs-2.14/header-attrs.js","resources_files/highlight-6.2.0/default.css","resources_files/highlight-6.2.0/highlight.js","resources_files/highlight-6.2.0/LICENSE","resources_files/highlight-6.2.0/textmate.css","resources_files/htmlwidgets-1.5.4/htmlwidgets.js","resources_files/jquery-3.6.0/jquery-3.6.0.js","resources_files/jquery-3.6.0/jquery-3.6.0.min.js","resources_files/jquery-3.6.0/jquery-3.6.0.min.map","resources_files/popper-2.6.0/popper.min.js","resources_files/profvis-0.3.6.9000/profvis.css","resources_files/profvis-0.3.6.9000/profvis.js","resources_files/profvis-0.3.6.9000/scroll.js","resources_files/profvis-binding-0.3.7/profvis.js","resources_files/tippy-6.2.7/tippy-bundle.umd.min.js","resources_files/tippy-6.2.7/tippy-light-border.css","resources_files/tippy-6.2.7/tippy.css","resources_files/tippy-6.2.7/tippy.umd.min.js","resources_files/webcomponents-2.0.0/webcomponents.js"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
    font-size: 100%;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme
    $('code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        const citeChild = $(this).children()[0]
        // Do not process if @xyz has been used without escaping and without bibliography activated
        // https://github.com/rstudio/distill/issues/466
        if (citeChild === undefined) return true

        if (citeChild.nodeName == "D-FOOTNOTE") {
          var fn = citeChild
          $(this).html(fn.shadowRoot.querySelector("sup"))
          $(this).id = fn.id
          fn.remove()
        }
        var refs = $(this).attr('data-cites').split(" ");
        var refHtml = refs.map(function(ref) {
          // Could use CSS.escape too here, we insure backward compatibility in navigator
          return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
        }).join("\n");
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // fix footnotes in tables (#411)
      // replacing broken distill.pub feature
      $('table d-footnote').each(function() {
        // we replace internal showAtNode methode which is triggered when hovering a footnote
        this.hoverBox.showAtNode = function(node) {
          // ported from https://github.com/distillpub/template/pull/105/files
          calcOffset = function(elem) {
              let x = elem.offsetLeft;
              let y = elem.offsetTop;
              // Traverse upwards until an `absolute` element is found or `elem`
              // becomes null.
              while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                  x += elem.offsetLeft;
                  y += elem.offsetTop;
              }

              return { left: x, top: y };
          }
          // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
          const bbox = node.getBoundingClientRect();
          const offset = calcOffset(node);
          this.show([offset.left + bbox.width, offset.top + bbox.height]);
        }
      })

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      // ignore leaflet img layers (#106)
      figures = figures.filter(':not(img[class*="leaflet"])')
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="resources_files/header-attrs-2.14/header-attrs.js"></script>
  <script src="resources_files/htmlwidgets-1.5.4/htmlwidgets.js"></script>
  <script src="resources_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="resources_files/d3-3.5.6/d3.min.js"></script>
  <link href="resources_files/profvis-0.3.6.9000/profvis.css" rel="stylesheet" />
  <script src="resources_files/profvis-0.3.6.9000/profvis.js"></script>
  <script src="resources_files/profvis-0.3.6.9000/scroll.js"></script>
  <link href="resources_files/highlight-6.2.0/textmate.css" rel="stylesheet" />
  <script src="resources_files/highlight-6.2.0/highlight.js"></script>
  <script src="resources_files/profvis-binding-0.3.7/profvis.js"></script>
  <script src="resources_files/popper-2.6.0/popper.min.js"></script>
  <link href="resources_files/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="resources_files/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="resources_files/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="resources_files/anchor-4.2.2/anchor.min.js"></script>
  <script src="resources_files/bowser-1.9.3/bowser.min.js"></script>
  <script src="resources_files/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="resources_files/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"Class wrap up: Data analysis, tips and resources","authors":[{"author":"Rui Fu, Kent Riemondy","authorURL":"#","affiliation":"&nbsp;","affiliationURL":"#","orcidID":""}],"publishedDate":"2022-12-05T00:00:00.000-07:00","citationText":"Riemondy, 2022"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Class wrap up: Data analysis, tips and resources</h1>
<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->

</div>

<div class="d-byline">
  Rui Fu, Kent Riemondy  
  
<br/>2022-12-05
</div>

<div class="d-article">
<div class="d-contents d-contents-float">
<nav class="l-text toc figcaption" id="TOC">
<h3>Contents</h3>
<ul>
<li><a href="#rmarkdown" id="toc-rmarkdown">Rmarkdown</a>
<ul>
<li><a href="#caching" id="toc-caching">Caching</a></li>
<li><a href="#better-save-and-load-.rds-files" id="toc-better-save-and-load-.rds-files">Better, save and load .rds files</a></li>
</ul></li>
<li><a href="#here-folder-structure-management" id="toc-here-folder-structure-management"><code>here</code>, folder structure management</a></li>
<li><a href="#styler-clean-up-code-readability" id="toc-styler-clean-up-code-readability"><code>styler</code>, clean up code readability</a></li>
<li><a href="#reproducibility" id="toc-reproducibility">Reproducibility</a>
<ul>
<li><a href="#print-environment-information-with-sessioninfo" id="toc-print-environment-information-with-sessioninfo">Print environment information with sessionInfo()</a></li>
<li><a href="#use-renv-to-manage-package-dependencies" id="toc-use-renv-to-manage-package-dependencies">Use renv to manage package dependencies</a></li>
</ul></li>
<li><a href="#benchmarking-with-microbenchmark-and-profvis" id="toc-benchmarking-with-microbenchmark-and-profvis">Benchmarking, with <code>microbenchmark</code> and <code>profvis</code></a></li>
<li><a href="#debugging-r-code" id="toc-debugging-r-code">Debugging R code</a>
<ul>
<li><a href="#look-at-the-call-stack-with-traceback" id="toc-look-at-the-call-stack-with-traceback">Look at the call stack with traceback()</a></li>
</ul></li>
<li><a href="#building-your-own-r-package" id="toc-building-your-own-r-package">Building your own R package</a></li>
<li><a href="#shiny-interactive-web-app-for-data-exploration" id="toc-shiny-interactive-web-app-for-data-exploration"><code>shiny</code>, interactive web app for data exploration</a></li>
<li><a href="#parsing-specific-types-of-formats" id="toc-parsing-specific-types-of-formats">Parsing specific types of formats:</a>
<ul>
<li><a href="#json" id="toc-json">JSON</a></li>
<li><a href="#genomics-data-fasta-fastq-vcf-bam-bed-bigwig" id="toc-genomics-data-fasta-fastq-vcf-bam-bed-bigwig">Genomics data (fasta, fastq, vcf, bam, bed, bigwig)</a></li>
</ul></li>
<li><a href="#using-r-on-the-command-line" id="toc-using-r-on-the-command-line">Using R on the command-line</a></li>
<li><a href="#git-and-github" id="toc-git-and-github">Git and Github</a>
<ul>
<li><a href="#put-your-code-on-github" id="toc-put-your-code-on-github">Put your code on GitHub</a></li>
<li><a href="#example-repos-rbi" id="toc-example-repos-rbi">Example repos (RBI)</a></li>
<li><a href="#asking-for-help-with-other-packages-on-github" id="toc-asking-for-help-with-other-packages-on-github">Asking for help with other packages on GitHub</a></li>
</ul></li>
<li><a href="#finding-useful-packages" id="toc-finding-useful-packages">Finding useful packages</a></li>
<li><a href="#bioconductor" id="toc-bioconductor">Bioconductor</a></li>
<li><a href="#finding-help-online" id="toc-finding-help-online">Finding help online</a></li>
<li><a href="#cheat-sheets" id="toc-cheat-sheets">Cheat sheets</a></li>
<li><a href="#offline-help" id="toc-offline-help">Offline help</a></li>
<li><a href="#sometimes-code-is-just-broken" id="toc-sometimes-code-is-just-broken">Sometimes code is just broken</a></li>
<li><a href="#additional-resources" id="toc-additional-resources">Additional Resources</a>
<ul>
<li><a href="#generaldata-science" id="toc-generaldata-science">General/Data science</a></li>
<li><a href="#genomics" id="toc-genomics">Genomics</a></li>
<li><a href="#a-meta-list-of-r-resources" id="toc-a-meta-list-of-r-resources">A meta-list of R resources</a></li>
<li><a href="#writing-high-performance-r-functions-with-r-c" id="toc-writing-high-performance-r-functions-with-r-c">Writing high-performance R functions with R + C++</a></li>
</ul></li>
</ul>
</nav>
</div>
<h2 id="rmarkdown">Rmarkdown</h2>
<p>Read the <a href="https://bookdown.org/yihui/rmarkdown/">Guide to RMarkdown</a> for an exhaustive description of the various formats and options for using RMarkdown documents. Note that HTML for this class were all made from Rmd, using the <a href="https://rstudio.github.io/distill/">distill blog format</a></p>
<h3 id="caching">Caching</h3>
<p>You can speed up knitting of your Rmds by using caching to store the results from each chunk, instead of rerunning them each time. Note that if you modify the code chunk, previous caching is ignored.</p>
<p>For each chunk, set <code>{r, cache = TRUE}</code></p>
<p>Or the option can be set globally at top of the document. Like this:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span>opts_chunk<span class="sc">$</span><span class="fu">set</span>(<span class="at">cache =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<h3 id="better-save-and-load-.rds-files">Better, save and load .rds files</h3>
<p>Run once, save, and load instead of rerunning resource intensive parts.</p>
<p>If you have non-deterministic functions, such as <code>kmeans</code>, remember to <code>set.seed</code>, or save and load result objects.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(<span class="st">&quot;long-step.rds&quot;</span>)){</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  ...</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  code or a script</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">source</span>(<span class="st">&quot;path/to/script.R&quot;</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  ...</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(obj, <span class="st">&quot;long-step.rds&quot;</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  obj <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;long-step.rds&quot;</span>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<h2 id="here-folder-structure-management"><code>here</code>, folder structure management</h2>
<p><a href="https://github.com/jennybc/here_here" class="uri">https://github.com/jennybc/here_here</a></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>here</span><span class='fu'>::</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='op'>)</span> <span class='co'># always points to top-level of current project</span></span>
<span></span>
<span><span class='fu'>here</span><span class='fu'>::</span><span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"_posts"</span>, <span class='st'>"2022-10-03-install-r"</span>, <span class='st'>"install-r.Rmd"</span><span class='op'>)</span> <span class='co'># never confused about folder structure</span></span></code></pre>
</div>
</div>
<h2 id="styler-clean-up-code-readability"><code>styler</code>, clean up code readability</h2>
<p>Refer to this <a href="http://adv-r.had.co.nz/Style.html">style guide</a> often, so you donâ€™t have to go back to make the code readable/publishable later.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>styler</span><span class='fu'>::</span><span class='fu'><a href='https://styler.r-lib.org/reference/style_text.html'>style_text</a></span><span class='op'>(</span><span class='st'>"</span></span>
<span><span class='st'>my_fun &lt;- function(x, </span></span>
<span><span class='st'>y, </span></span>
<span><span class='st'>z) {</span></span>
<span><span class='st'>  x+ z</span></span>
<span><span class='st'>}</span></span>
<span><span class='st'>                   "</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>#&gt; 
#&gt; my_fun &lt;- function(x,
#&gt;                    y,
#&gt;                    z) {
#&gt;   x + z
#&gt; }</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span></span>
<span><span class='co'># styler::style_file   # for an entire file</span></span>
<span><span class='co'># styler::style_dir    # directory </span></span></code></pre>
</div>
</div>
<h2 id="reproducibility">Reproducibility</h2>
<h3 id="print-environment-information-with-sessioninfo">Print environment information with sessionInfo()</h3>
<p>Itâ€™s very helpful to have a record of which packages were used in an analysis. One approach is to print the <code>sessionInfo()</code>.</p>
<details>
<summary>
Show session info
</summary>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/utils/sessionInfo.html'>sessionInfo</a></span><span class='op'>(</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>#&gt; R version 4.2.0 (2022-04-22)
#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)
#&gt; Running under: macOS Big Sur/Monterey 10.16
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib
#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib
#&gt; 
#&gt; locale:
#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#&gt; 
#&gt; attached base packages:
#&gt; [1] stats     graphics  grDevices utils     datasets  methods  
#&gt; [7] base     
#&gt; 
#&gt; other attached packages:
#&gt;  [1] here_1.0.1      forcats_0.5.1   stringr_1.4.1   dplyr_1.0.10   
#&gt;  [5] purrr_0.3.5     readr_2.1.2     tidyr_1.2.0     tibble_3.1.8   
#&gt;  [9] ggplot2_3.3.6   tidyverse_1.3.1
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;  [1] lubridate_1.8.0   assertthat_0.2.1  rprojroot_2.0.3  
#&gt;  [4] digest_0.6.30     utf8_1.2.2        prettycode_1.1.0 
#&gt;  [7] R6_2.5.1          cellranger_1.1.0  backports_1.4.1  
#&gt; [10] reprex_2.0.1      evaluate_0.16     httr_1.4.4       
#&gt; [13] pillar_1.8.1      rlang_1.0.6       readxl_1.4.0     
#&gt; [16] rstudioapi_0.13   jquerylib_0.1.4   R.utils_2.12.0   
#&gt; [19] R.oo_1.25.0       rmarkdown_2.14    styler_1.7.0     
#&gt; [22] munsell_0.5.0     broom_0.8.0       compiler_4.2.0   
#&gt; [25] modelr_0.1.8      xfun_0.32         pkgconfig_2.0.3  
#&gt; [28] htmltools_0.5.2   downlit_0.4.2     tidyselect_1.2.0 
#&gt; [31] fansi_1.0.3       crayon_1.5.2      tzdb_0.3.0       
#&gt; [34] dbplyr_2.2.1      withr_2.5.0       R.methodsS3_1.8.2
#&gt; [37] grid_4.2.0        jsonlite_1.8.3    gtable_0.3.0     
#&gt; [40] lifecycle_1.0.3   DBI_1.1.3         magrittr_2.0.3   
#&gt; [43] scales_1.2.0      cli_3.4.1         stringi_1.7.8    
#&gt; [46] cachem_1.0.6      fs_1.5.2          xml2_1.3.3       
#&gt; [49] bslib_0.3.1       ellipsis_0.3.2    generics_0.1.3   
#&gt; [52] vctrs_0.4.1       distill_1.5       tools_4.2.0      
#&gt; [55] R.cache_0.15.0    glue_1.6.2        hms_1.1.2        
#&gt; [58] fastmap_1.1.0     yaml_2.3.6        colorspace_2.0-3 
#&gt; [61] rvest_1.0.2       memoise_2.0.1     knitr_1.39       
#&gt; [64] haven_2.5.0       sass_0.4.1</code></pre>
</div>
<p>See also the <code>sessioninfo</code> package, which provide more details:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>sessioninfo</span><span class='fu'>::</span><span class='fu'><a href='https://r-lib.github.io/sessioninfo/reference/session_info.html'>session_info</a></span><span class='op'>(</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>#&gt; â”€ Session info â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
#&gt;  setting  value
#&gt;  version  R version 4.2.0 (2022-04-22)
#&gt;  os       macOS Big Sur/Monterey 10.16
#&gt;  system   x86_64, darwin17.0
#&gt;  ui       X11
#&gt;  language (EN)
#&gt;  collate  en_US.UTF-8
#&gt;  ctype    en_US.UTF-8
#&gt;  tz       America/Denver
#&gt;  date     2022-12-16
#&gt;  pandoc   2.19.2 @ /Applications/RStudio.app/Contents/MacOS/quarto/bin/tools/ (via rmarkdown)
#&gt; 
#&gt; â”€ Packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
#&gt;  package     * version date (UTC) lib source
#&gt;  assertthat    0.2.1   2019-03-21 [1] CRAN (R 4.2.0)
#&gt;  backports     1.4.1   2021-12-13 [1] CRAN (R 4.2.0)
#&gt;  broom         0.8.0   2022-04-13 [1] CRAN (R 4.2.0)
#&gt;  bslib         0.3.1   2021-10-06 [1] CRAN (R 4.2.0)
#&gt;  cachem        1.0.6   2021-08-19 [1] CRAN (R 4.2.0)
#&gt;  cellranger    1.1.0   2016-07-27 [1] CRAN (R 4.2.0)
#&gt;  cli           3.4.1   2022-09-23 [1] CRAN (R 4.2.0)
#&gt;  colorspace    2.0-3   2022-02-21 [1] CRAN (R 4.2.0)
#&gt;  crayon        1.5.2   2022-09-29 [1] CRAN (R 4.2.0)
#&gt;  DBI           1.1.3   2022-06-18 [1] CRAN (R 4.2.0)
#&gt;  dbplyr        2.2.1   2022-06-27 [1] CRAN (R 4.2.0)
#&gt;  digest        0.6.30  2022-10-18 [1] CRAN (R 4.2.0)
#&gt;  distill       1.5     2022-09-07 [1] CRAN (R 4.2.0)
#&gt;  downlit       0.4.2   2022-07-05 [1] CRAN (R 4.2.0)
#&gt;  dplyr       * 1.0.10  2022-09-01 [1] CRAN (R 4.2.0)
#&gt;  ellipsis      0.3.2   2021-04-29 [1] CRAN (R 4.2.0)
#&gt;  evaluate      0.16    2022-08-09 [1] CRAN (R 4.2.0)
#&gt;  fansi         1.0.3   2022-03-24 [1] CRAN (R 4.2.0)
#&gt;  fastmap       1.1.0   2021-01-25 [1] CRAN (R 4.2.0)
#&gt;  forcats     * 0.5.1   2021-01-27 [1] CRAN (R 4.2.0)
#&gt;  fs            1.5.2   2021-12-08 [1] CRAN (R 4.2.0)
#&gt;  generics      0.1.3   2022-07-05 [1] CRAN (R 4.2.0)
#&gt;  ggplot2     * 3.3.6   2022-05-03 [1] CRAN (R 4.2.0)
#&gt;  glue          1.6.2   2022-02-24 [1] CRAN (R 4.2.0)
#&gt;  gtable        0.3.0   2019-03-25 [1] CRAN (R 4.2.0)
#&gt;  haven         2.5.0   2022-04-15 [1] CRAN (R 4.2.0)
#&gt;  here        * 1.0.1   2020-12-13 [1] CRAN (R 4.2.0)
#&gt;  hms           1.1.2   2022-08-19 [1] CRAN (R 4.2.0)
#&gt;  htmltools     0.5.2   2021-08-25 [1] CRAN (R 4.2.0)
#&gt;  httr          1.4.4   2022-08-17 [1] CRAN (R 4.2.0)
#&gt;  jquerylib     0.1.4   2021-04-26 [1] CRAN (R 4.2.0)
#&gt;  jsonlite      1.8.3   2022-10-21 [1] CRAN (R 4.2.0)
#&gt;  knitr         1.39    2022-04-26 [1] CRAN (R 4.2.0)
#&gt;  lifecycle     1.0.3   2022-10-07 [1] CRAN (R 4.2.0)
#&gt;  lubridate     1.8.0   2021-10-07 [1] CRAN (R 4.2.0)
#&gt;  magrittr      2.0.3   2022-03-30 [1] CRAN (R 4.2.0)
#&gt;  memoise       2.0.1   2021-11-26 [1] CRAN (R 4.2.0)
#&gt;  modelr        0.1.8   2020-05-19 [1] CRAN (R 4.2.0)
#&gt;  munsell       0.5.0   2018-06-12 [1] CRAN (R 4.2.0)
#&gt;  pillar        1.8.1   2022-08-19 [1] CRAN (R 4.2.0)
#&gt;  pkgconfig     2.0.3   2019-09-22 [1] CRAN (R 4.2.0)
#&gt;  prettycode    1.1.0   2019-12-16 [1] CRAN (R 4.2.0)
#&gt;  purrr       * 0.3.5   2022-10-06 [1] CRAN (R 4.2.0)
#&gt;  R.cache       0.15.0  2021-04-30 [1] CRAN (R 4.2.0)
#&gt;  R.methodsS3   1.8.2   2022-06-13 [1] CRAN (R 4.2.0)
#&gt;  R.oo          1.25.0  2022-06-12 [1] CRAN (R 4.2.0)
#&gt;  R.utils       2.12.0  2022-06-28 [1] CRAN (R 4.2.0)
#&gt;  R6            2.5.1   2021-08-19 [1] CRAN (R 4.2.0)
#&gt;  readr       * 2.1.2   2022-01-30 [1] CRAN (R 4.2.0)
#&gt;  readxl        1.4.0   2022-03-28 [1] CRAN (R 4.2.0)
#&gt;  reprex        2.0.1   2021-08-05 [1] CRAN (R 4.2.0)
#&gt;  rlang         1.0.6   2022-09-24 [1] CRAN (R 4.2.0)
#&gt;  rmarkdown     2.14    2022-04-25 [1] CRAN (R 4.2.0)
#&gt;  rprojroot     2.0.3   2022-04-02 [1] CRAN (R 4.2.0)
#&gt;  rstudioapi    0.13    2020-11-12 [1] CRAN (R 4.2.0)
#&gt;  rvest         1.0.2   2021-10-16 [1] CRAN (R 4.2.0)
#&gt;  sass          0.4.1   2022-03-23 [1] CRAN (R 4.2.0)
#&gt;  scales        1.2.0   2022-04-13 [1] CRAN (R 4.2.0)
#&gt;  sessioninfo   1.2.2   2021-12-06 [1] CRAN (R 4.2.0)
#&gt;  stringi       1.7.8   2022-07-11 [1] CRAN (R 4.2.0)
#&gt;  stringr     * 1.4.1   2022-08-20 [1] CRAN (R 4.2.0)
#&gt;  styler        1.7.0   2022-03-13 [1] CRAN (R 4.2.0)
#&gt;  tibble      * 3.1.8   2022-07-22 [1] CRAN (R 4.2.0)
#&gt;  tidyr       * 1.2.0   2022-02-01 [1] CRAN (R 4.2.0)
#&gt;  tidyselect    1.2.0   2022-10-10 [1] CRAN (R 4.2.0)
#&gt;  tidyverse   * 1.3.1   2021-04-15 [1] CRAN (R 4.2.0)
#&gt;  tzdb          0.3.0   2022-03-28 [1] CRAN (R 4.2.0)
#&gt;  utf8          1.2.2   2021-07-24 [1] CRAN (R 4.2.0)
#&gt;  vctrs         0.4.1   2022-04-13 [1] CRAN (R 4.2.0)
#&gt;  withr         2.5.0   2022-03-03 [1] CRAN (R 4.2.0)
#&gt;  xfun          0.32    2022-08-10 [1] CRAN (R 4.2.0)
#&gt;  xml2          1.3.3   2021-11-30 [1] CRAN (R 4.2.0)
#&gt;  yaml          2.3.6   2022-10-18 [1] CRAN (R 4.2.0)
#&gt; 
#&gt;  [1] /Library/Frameworks/R.framework/Versions/4.2/Resources/library
#&gt; 
#&gt; â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</code></pre>
</div>
</details>
<h3 id="use-renv-to-manage-package-dependencies">Use renv to manage package dependencies</h3>
<p>The <a href="https://rstudio.github.io/renv/articles/renv.html"><code>renv</code></a> package allows you to have a separate set of R packages for each project. It also can record and restore the set of R pacakges used in a project. This is very helpful when you need to return to a project months (or years) later and want to have the same set of packages. It also makes it easier to share your packages with collaborators.</p>
<p>See also:</p>
<ul>
<li><p><a href="https://docs.conda.io/en/latest/">conda</a> for managing various command line programs (python, R, c, etc.)</p></li>
<li><p><a href="https://www.docker.com/">docker</a> for generating a fully reproducible operating system environment.</p></li>
</ul>
<h2 id="benchmarking-with-microbenchmark-and-profvis">Benchmarking, with <code>microbenchmark</code> and <code>profvis</code></h2>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># example, compare base and readr csv reading functions</span></span>
<span><span class='va'>path_to_file</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://here.r-lib.org//reference/here.html'>here</a></span><span class='op'>(</span><span class='st'>"data/class3/dmel_peptides_lifecycle.csv.gz"</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>res</span> <span class='op'>&lt;-</span> <span class='fu'>microbenchmark</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/microbenchmark/man/microbenchmark.html'>microbenchmark</a></span><span class='op'>(</span></span>
<span>  base <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.csv</a></span><span class='op'>(</span><span class='va'>path_to_file</span><span class='op'>)</span>,</span>
<span>  readr <span class='op'>=</span> <span class='fu'>readr</span><span class='fu'>::</span><span class='fu'><a href='https://readr.tidyverse.org/reference/read_delim.html'>read_csv</a></span><span class='op'>(</span><span class='va'>path_to_file</span><span class='op'>)</span>,</span>
<span>  times <span class='op'>=</span> <span class='fl'>5</span></span>
<span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>res</span>, signif <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>#&gt; Unit: milliseconds
#&gt;   expr  min   lq mean median   uq  max neval
#&gt;   base 3400 3400 3500   3400 3600 3700     5
#&gt;  readr  300  300  390    310  410  650     5</code></pre>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># example, looking at each step of a script</span></span>
<span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://tidyverse.tidyverse.org'>tidyverse</a></span><span class='op'>)</span></span>
<span><span class='va'>p</span> <span class='op'>&lt;-</span> <span class='fu'>profvis</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/profvis/man/profvis.html'>profvis</a></span><span class='op'>(</span><span class='op'>{</span></span>
<span></span>
<span>  <span class='va'>df</span> <span class='op'>&lt;-</span> <span class='fu'>readr</span><span class='fu'>::</span><span class='fu'><a href='https://readr.tidyverse.org/reference/read_delim.html'>read_csv</a></span><span class='op'>(</span><span class='va'>path_to_file</span><span class='op'>)</span></span>
<span>  <span class='va'>df</span> <span class='op'>&lt;-</span> <span class='va'>df</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span></span>
<span>    <span class='fu'><a href='https://dplyr.tidyverse.org/reference/filter.html'>filter</a></span><span class='op'>(</span><span class='va'>e02_4</span> <span class='op'>&lt;</span> <span class='fl'>50</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span></span>
<span>    <span class='fu'><a href='https://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span><span class='op'>(</span>sequence <span class='op'>=</span> <span class='fu'><a href='https://stringr.tidyverse.org/reference/case.html'>str_to_lower</a></span><span class='op'>(</span><span class='va'>Sequence</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='op'>}</span><span class='op'>)</span></span>
<span><span class='va'>p</span></span></code></pre>
</div>
<div id="htmlwidget-19a72b88df8477c4c18e" style="width:100%;height:600px;" class="profvis html-widget"></div>
<script type="application/json" data-for="htmlwidget-19a72b88df8477c4c18e">{"x":{"message":{"prof":{"time":[1,1,2,2,3,3,4,4,5,5,6,6,7,7,8,8,9,9,10,10,11,11,12,12,13,13,14,14,15,15,16,16,17,17,18,18,19,19,20,20,21,22,23,23,24,24,25,25,26,26,27,27,28,28,29,29,30,30,31,31,32,32,33,33,34,34,35,35,36,36,37,37,38,38,39,39,40,40,41,41,42,42,43,43,44,44,45,45,46,46,47,47,48,48,49,49,50,50,51,51,52,52,53,53,54,54,55,55,56,56,57,57,58,58,59,59,60,60,61,61,62,62,63,63,64,64,65,65,66,66,67,67,67,67,67,67,67,67,67,67,67,67,67,68,68,68,68,68,68,68,68,68,68,68,68,69,69,69,69,69,69,69,69,69,69,69,69,69,69,69,69,69,70,70,70,70,71,71,72,72,72],"depth":[2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,1,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,13,12,11,10,9,8,7,6,5,4,3,2,1,12,11,10,9,8,7,6,5,4,3,2,1,17,16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1,4,3,2,1,2,1,3,2,1],"label":[".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render","<Anonymous>","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render","<Anonymous>","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render","rmarkdown::render","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render",".Call","rmarkdown::render","unique.default","c.numeric_version","Ops.numeric_version",".rlang_cli_compat",".rlang_cli_format_fallback","cnd_message_info","type","cli_block","show_dims","show_col_types","vroom::vroom","readr::read_csv","rmarkdown::render","lapply","packageDescription","utils::packageVersion",".rlang_cli_compat",".rlang_cli_format_fallback","cnd_message_info","type","cli_block","show_col_types","vroom::vroom","readr::read_csv","rmarkdown::render","cmp","h","tryInline","cmpCall","cmp","h","tryInline","cmpCall","cmp","h","tryInline","cmpCall","cmp","genCode","cmpfun","compiler:::tryCmpfun","rmarkdown::render","vec_slice","dplyr_row_slice.data.frame","filter.data.frame","rmarkdown::render","lazyLoadDBfetch","rmarkdown::render","stri_trans_tolower","str_to_lower","rmarkdown::render"],"filenum":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,1,null,null,null,null,null,null,null,null,null,null,null,1,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"linenum":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,5,null,null,null,null,null,null,null,null,null,null,null,5,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"memalloc":[19.1502227783203,19.1502227783203,20.0257034301758,20.0257034301758,20.9011840820312,20.9011840820312,21.7766647338867,21.7766647338867,22.6521453857422,22.6521453857422,23.4025573730469,23.4025573730469,24.2780380249023,24.2780380249023,25.1535186767578,25.1535186767578,26.0289993286133,26.0289993286133,26.9044799804688,26.9044799804688,27.7799606323242,27.7799606323242,28.6554412841797,28.6554412841797,29.6559906005859,29.6559906005859,30.4064025878906,30.4064025878906,31.2818832397461,31.2818832397461,32.0487442016602,32.0487442016602,32.2090530395508,32.2090530395508,32.3678359985352,32.3678359985352,32.5335006713867,32.5335006713867,32.6971588134766,32.6971588134766,33.7305374145508,34.6737289428711,35.1298904418945,35.1298904418945,35.5629119873047,35.5629119873047,36.428955078125,36.428955078125,36.8619766235352,36.8619766235352,37.7280197143555,37.7280197143555,38.1610412597656,38.1610412597656,39.0270843505859,39.0270843505859,39.4601058959961,39.4601058959961,40.3261489868164,40.3261489868164,41.1921920776367,41.1921920776367,41.6252136230469,41.6252136230469,42.4912567138672,42.4912567138672,43.3572998046875,43.3572998046875,43.7903213500977,43.7903213500977,44.656364440918,44.656364440918,45.0893859863281,45.0893859863281,45.9554290771484,45.9554290771484,46.3884506225586,46.3884506225586,47.2544937133789,47.2544937133789,47.6875152587891,47.6875152587891,48.5535583496094,48.5535583496094,49.4196014404297,49.4196014404297,49.8526229858398,49.8526229858398,50.7186660766602,50.7186660766602,51.5847091674805,51.5847091674805,52.0177307128906,52.0177307128906,52.8837738037109,52.8837738037109,53.3167953491211,53.3167953491211,54.1828384399414,54.1828384399414,54.6158599853516,54.6158599853516,55.4819030761719,55.4819030761719,55.914924621582,55.914924621582,56.7809677124023,56.7809677124023,57.2139892578125,57.2139892578125,58.0800323486328,58.0800323486328,58.513053894043,58.513053894043,59.3790969848633,59.3790969848633,60.2451400756836,60.2451400756836,60.6781616210938,60.6781616210938,61.5442047119141,61.5442047119141,62.4102478027344,62.4102478027344,62.8432693481445,62.8432693481445,63.7093124389648,63.7093124389648,64.142333984375,64.142333984375,64.3235092163086,64.3235092163086,64.3235092163086,64.3235092163086,64.3235092163086,64.3235092163086,64.3235092163086,64.3235092163086,64.3235092163086,64.3235092163086,64.3235092163086,64.3235092163086,64.3235092163086,64.5690612792969,64.5690612792969,64.5690612792969,64.5690612792969,64.5690612792969,64.5690612792969,64.5690612792969,64.5690612792969,64.5690612792969,64.5690612792969,64.5690612792969,64.5690612792969,65.3418655395508,65.3418655395508,65.3418655395508,65.3418655395508,65.3418655395508,65.3418655395508,65.3418655395508,65.3418655395508,65.3418655395508,65.3418655395508,65.3418655395508,65.3418655395508,65.3418655395508,65.3418655395508,65.3418655395508,65.3418655395508,65.3418655395508,75.5080642700195,75.5080642700195,75.5080642700195,75.5080642700195,76.2305603027344,76.2305603027344,77.3224945068359,77.3224945068359,77.3224945068359],"meminc":[0,0,0.875480651855469,0,0.875480651855469,0,0.875480651855469,0,0.875480651855469,0,0.750411987304688,0,0.875480651855469,0,0.875480651855469,0,0.875480651855469,0,0.875480651855469,0,0.875480651855469,0,0.875480651855469,0,1.00054931640625,0,0.750411987304688,0,0.875480651855469,0,0.766860961914062,0,0.160308837890625,0,0.158782958984375,0,0.165664672851562,0,0.163658142089844,0,1.03337860107422,0.943191528320312,0.456161499023438,0,0.433021545410156,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.866043090820312,0,0.433021545410156,0,0.866043090820312,0,0.433021545410156,0,0.181175231933594,0,0,0,0,0,0,0,0,0,0,0,0,0.245552062988281,0,0,0,0,0,0,0,0,0,0,0,0.772804260253906,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,10.1661987304688,0,0,0,0.722496032714844,0,1.09193420410156,0,0],"filename":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"<expr>",null,null,null,null,null,null,null,null,null,null,null,"<expr>",null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null]},"interval":10,"files":[{"filename":"<expr>","content":"# example, looking at each step of a script\nlibrary(tidyverse)\np <- profvis::profvis({\n\n  df <- readr::read_csv(path_to_file)\n  df <- df %>%\n    filter(e02_4 < 50) %>%\n    mutate(sequence = str_to_lower(Sequence))\n})\np","normpath":"<expr>"}],"prof_output":"/var/folders/r9/g3c47jrj40gc14d8qsqx7src0000gn/T//RtmpOVyrhD/file475876a19610.prof","highlight":{"output":["^output\\$"],"gc":["^<GC>$"],"stacktrace":["^\\.\\.stacktraceo(n|ff)\\.\\.$"]},"split":"h"}},"evals":[],"jsHooks":[]}</script>
</div>
<h2 id="debugging-r-code">Debugging R code</h2>
<p>R has a debugger built in. You can debug a function:</p>
<p>e.g.:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/debug.html'>debug</a></span><span class='op'>(</span><span class='va'>read_csv</span><span class='op'>)</span> <span class='co'># set a function to debug</span></span>
<span><span class='fu'><a href='https://readr.tidyverse.org/reference/read_delim.html'>read_csv</a></span><span class='op'>(</span><span class='va'>path_to_file</span><span class='op'>)</span> <span class='co'># will enter debug mode</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/debug.html'>undebug</a></span><span class='op'>(</span><span class='va'>read_csv</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>Rstudio has great support for debugging functions in Rscripts or in packages:</p>
<p><a href="https://support.posit.co/hc/en-us/articles/200713843-Debugging-R-code-with-the-RStudio-IDE" class="uri">https://support.posit.co/hc/en-us/articles/200713843-Debugging-R-code-with-the-RStudio-IDE</a></p>
<h3 id="look-at-the-call-stack-with-traceback">Look at the call stack with traceback()</h3>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>cool_function</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='op'>{</span></span>
<span>  </span>
<span>  <span class='va'>internal_function</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>y</span><span class='op'>)</span> <span class='op'>{</span></span>
<span>    </span>
<span>    <span class='va'>hard_to_find</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>z</span><span class='op'>)</span> <span class='op'>{</span></span>
<span>      <span class='st'>"doesn't work here"</span> <span class='op'>+</span> <span class='fl'>10</span></span>
<span>    <span class='op'>}</span></span>
<span>    </span>
<span>    <span class='fu'>hard_to_find</span><span class='op'>(</span><span class='op'>)</span></span>
<span>  <span class='op'>}</span></span>
<span>  <span class='fu'>internal_function</span><span class='op'>(</span><span class='op'>)</span></span>
<span><span class='op'>}</span></span>
<span></span>
<span><span class='fu'>cool_function</span><span class='op'>(</span><span class='fl'>1</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/traceback.html'>traceback</a></span><span class='op'>(</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p><img src="https://github.com/rnabioco/bmsc-7810-pbda/raw/main/_posts/2022-12-05-a-collection-of-r-resources/img/traceback.png" /></p>
<h2 id="building-your-own-r-package">Building your own R package</h2>
<p>Itâ€™s surprisingly easy, particularly with Rstudio, to write your own R package to store your code. Putting your code in a package makes it much easier to debug, document, add tests, and distribute your code.</p>
<ul>
<li><a href="https://r-pkgs.org/" class="uri">https://r-pkgs.org/</a></li>
<li><a href="https://hilaryparker.com/2014/04/29/writing-an-r-package-from-scratch/" class="uri">https://hilaryparker.com/2014/04/29/writing-an-r-package-from-scratch/</a></li>
</ul>
<h2 id="shiny-interactive-web-app-for-data-exploration"><code>shiny</code>, interactive web app for data exploration</h2>
<p>Making an interactive interface to data and plotting is easy in R. Examples and corresponding code can be found at <a href="https://shiny.rstudio.com/gallery/" class="uri">https://shiny.rstudio.com/gallery/</a>.</p>
<h2 id="parsing-specific-types-of-formats">Parsing specific types of formats:</h2>
<p>There are generally packages built to read and write from most formats. Some examples:</p>
<h3 id="json">JSON</h3>
<p>Check out <a href="https://cran.r-project.org/web/packages/jsonlite/vignettes/json-aaquickstart.html">jsonlite</a></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://arxiv.org/abs/1403.2805'>jsonlite</a></span><span class='op'>)</span></span>
<span><span class='va'>json_file</span> <span class='op'>&lt;-</span> <span class='st'>"http://api.worldbank.org/country?per_page=10&amp;region=OED&amp;lendingtype=LNX&amp;format=json"</span></span>
<span><span class='va'>worldbank_data</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/jsonlite/man/fromJSON.html'>fromJSON</a></span><span class='op'>(</span><span class='va'>json_file</span>, flatten<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span></span>
<span><span class='va'>worldbank_data</span></span></code></pre>
</div>
<pre><code>#&gt; [[1]]
#&gt; [[1]]$page
#&gt; [1] 1
#&gt; 
#&gt; [[1]]$pages
#&gt; [1] 4
#&gt; 
#&gt; [[1]]$per_page
#&gt; [1] &quot;10&quot;
#&gt; 
#&gt; [[1]]$total
#&gt; [1] 32
#&gt; 
#&gt; 
#&gt; [[2]]
#&gt;     id iso2Code        name capitalCity longitude latitude region.id
#&gt; 1  AUS       AU   Australia    Canberra   149.129  -35.282       EAS
#&gt; 2  AUT       AT     Austria      Vienna   16.3798  48.2201       ECS
#&gt; 3  BEL       BE     Belgium    Brussels   4.36761  50.8371       ECS
#&gt; 4  CAN       CA      Canada      Ottawa  -75.6919  45.4215       NAC
#&gt; 5  CHE       CH Switzerland        Bern   7.44821   46.948       ECS
#&gt; 6  CZE       CZ     Czechia      Prague   14.4205  50.0878       ECS
#&gt; 7  DEU       DE     Germany      Berlin   13.4115  52.5235       ECS
#&gt; 8  DNK       DK     Denmark  Copenhagen   12.5681  55.6763       ECS
#&gt; 9  ESP       ES       Spain      Madrid  -3.70327  40.4167       ECS
#&gt; 10 EST       EE     Estonia     Tallinn   24.7586  59.4392       ECS
#&gt;    region.iso2code          region.value adminregion.id
#&gt; 1               Z4   East Asia &amp; Pacific               
#&gt; 2               Z7 Europe &amp; Central Asia               
#&gt; 3               Z7 Europe &amp; Central Asia               
#&gt; 4               XU         North America               
#&gt; 5               Z7 Europe &amp; Central Asia               
#&gt; 6               Z7 Europe &amp; Central Asia               
#&gt; 7               Z7 Europe &amp; Central Asia               
#&gt; 8               Z7 Europe &amp; Central Asia               
#&gt; 9               Z7 Europe &amp; Central Asia               
#&gt; 10              Z7 Europe &amp; Central Asia               
#&gt;    adminregion.iso2code adminregion.value incomeLevel.id
#&gt; 1                                                    HIC
#&gt; 2                                                    HIC
#&gt; 3                                                    HIC
#&gt; 4                                                    HIC
#&gt; 5                                                    HIC
#&gt; 6                                                    HIC
#&gt; 7                                                    HIC
#&gt; 8                                                    HIC
#&gt; 9                                                    HIC
#&gt; 10                                                   HIC
#&gt;    incomeLevel.iso2code incomeLevel.value lendingType.id
#&gt; 1                    XD       High income            LNX
#&gt; 2                    XD       High income            LNX
#&gt; 3                    XD       High income            LNX
#&gt; 4                    XD       High income            LNX
#&gt; 5                    XD       High income            LNX
#&gt; 6                    XD       High income            LNX
#&gt; 7                    XD       High income            LNX
#&gt; 8                    XD       High income            LNX
#&gt; 9                    XD       High income            LNX
#&gt; 10                   XD       High income            LNX
#&gt;    lendingType.iso2code lendingType.value
#&gt; 1                    XX    Not classified
#&gt; 2                    XX    Not classified
#&gt; 3                    XX    Not classified
#&gt; 4                    XX    Not classified
#&gt; 5                    XX    Not classified
#&gt; 6                    XX    Not classified
#&gt; 7                    XX    Not classified
#&gt; 8                    XX    Not classified
#&gt; 9                    XX    Not classified
#&gt; 10                   XX    Not classified</code></pre>
</div>
<h3 id="genomics-data-fasta-fastq-vcf-bam-bed-bigwig">Genomics data (fasta, fastq, vcf, bam, bed, bigwig)</h3>
<p>Check out <code>rtracklayer</code> and <code>Rsamtools</code>:</p>
<p>e.g.Â read a FASTA file into R:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://bioconductor.org/packages/Rsamtools'>Rsamtools</a></span><span class='op'>)</span></span>
<span></span>
<span><span class='co'># get path to test file included in a package</span></span>
<span><span class='va'>fasta_file</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>'extdata'</span>, <span class='st'>'ce2dict1.fa'</span>, package <span class='op'>=</span> <span class='st'>'Rsamtools'</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/pkg/Rsamtools/man/scanFa.html'>scanFa</a></span><span class='op'>(</span><span class='va'>fasta_file</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>#&gt; DNAStringSet object of length 5:
#&gt;     width seq                                     names               
#&gt; [1]    18 GCGAAACTAGGAGAGGCT                      pattern01
#&gt; [2]    25 CTGTTAGCTAATTTTAAAAATAAAT               pattern02
#&gt; [3]    24 ACTACCACCCAAATTTAGATATTC                pattern03
#&gt; [4]    24 AAATTTTTTTTGTTGCAAATTTGA                pattern04
#&gt; [5]    25 TCTTCTTGGCTTTGGTGGTACTTTT               pattern05</code></pre>
</div>
<h2 id="using-r-on-the-command-line">Using R on the command-line</h2>
<p>The command line can be accessed via the Terminal app on macOS, or using the windows subsystem for linux (WSL).</p>
<p>There command line is a place where you can run executable programs (a C, python, R, or whatever). Itâ€™s what using a computer looked like before the existence of a Graphical User Interface. It is impossible to conduct data analysis without gaining some experience with working on the command line.</p>
<p>R is an executable, and we can pull up an R console using:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<p>In Rmarkdown you can also include other languages including bash (which is a common language of the command line).You need to change the <code>r</code> to <code>bash</code> in the code chunk (or to python or other languages).</p>
<p>You can run simple commands by using <code>Rscript</code> or <code>R</code> with the <code>-e</code> option.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-e</span> <span class="st">&quot;print(&#39;hello&#39;)&quot;</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; R version 4.2.0 (2022-04-22) -- &quot;Vigorous Calisthenics&quot;</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Copyright (C) 2022 The R Foundation for Statistical Computing</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; R is free software and comes with ABSOLUTELY NO WARRANTY.</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; You are welcome to redistribute it under certain conditions.</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Type &#39;license()&#39; or &#39;licence()&#39; for distribution details.</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   Natural language support but running in an English locale</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; R is a collaborative project with many contributors.</span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Type &#39;contributors()&#39; for more information and</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; &#39;citation()&#39; on how to cite R or R packages in publications.</span></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Type &#39;demo()&#39; for some demos, &#39;help()&#39; for on-line help, or</span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; &#39;help.start()&#39; for an HTML browser interface to help.</span></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Type &#39;q()&#39; to quit R.</span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; &gt; print(&#39;hello&#39;)</span></span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;hello&quot;</span></span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; &gt; </span></span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; &gt;</span></span></code></pre></div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">&quot;print(&#39;hello&#39;)&quot;</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;hello&quot;</span></span></code></pre></div>
</div>
<p>Alternatively you can write a R script, which can be then called from Rscript. For example if we wrote an R script called <code>cool_function.R</code>.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/usr/bin/env Rscript  # allows calling with ./cool_function.R if executable</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>args <span class="ot">=</span> <span class="fu">commandArgs</span>(<span class="at">trailingOnly=</span><span class="cn">TRUE</span>) <span class="co"># collect command line arguments </span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(args) <span class="co"># args is a list e.g. argument1 argument2...</span></span></code></pre></div>
<p>We could call on the command line:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> path/to/cool_function.R argument1 argument2 ...</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co">#or</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="ex">path/to/cool_function.R</span> argument1 argument2 ...</span></code></pre></div>
<h2 id="git-and-github">Git and Github</h2>
<figure>
<img src="https://github.com/rnabioco/bmsc-7810-pbda/raw/main/_posts/2022-12-05-a-collection-of-r-resources/img/git-basics.png" alt="From https://jmcglone.com/guides/github-pages/" />
<figcaption aria-hidden="true">From <a href="https://jmcglone.com/guides/github-pages/" class="uri">https://jmcglone.com/guides/github-pages/</a></figcaption>
</figure>
<p>Git is a command line tool for version control, which allows us to:</p>
<ol type="1">
<li><p>rolling back code to a previous state if needed</p></li>
<li><p>branched development, tackling individual issues/tasks</p></li>
<li><p>collaboration</p></li>
</ol>
<figure>
<img src="https://github.com/rnabioco/bmsc-7810-pbda/raw/main/_posts/2022-12-05-a-collection-of-r-resources/img/github-flow.png" alt="From https://blog.programster.org/git-workflows" />
<figcaption aria-hidden="true">From <a href="https://blog.programster.org/git-workflows" class="uri">https://blog.programster.org/git-workflows</a></figcaption>
</figure>
<p>Git was first created by Linus Torvalds for coordinating development of Linux. Read this guide for <a href="https://git-scm.com/book/en/v2/Getting-Started-About-Version-Control">Getting started</a>
, checkout this <a href="https://learngitbranching.js.org/">interactive guide</a> and check out this <a href="https://happygitwithr.com">Tutorial</a> written from an R data analyst perspective.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for bioinformatics, get comfortable with command line too</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> status <span class="co"># list changes to tracked files</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> blame resources.Rmd    <span class="co"># see who contributed</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> commit <span class="at">-m</span> <span class="st">&quot;added something cool&quot;</span> <span class="co"># save state</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> push <span class="co"># push git to a git repository (e.g. github)</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> pull <span class="co"># pull changes from git repository</span></span></code></pre></div>
</div>
<p>This can be handled by Rstudio as well (new tab next to <code>Connections</code> and <code>Build</code>)</p>
<h3 id="put-your-code-on-github">Put your code on GitHub</h3>
<p>As you write more code, especially as functions and script pipelines, hosting and documenting them on GitHub is great way to make them portable and searchable. Even the free tier of GitHub accounts now has private repositories (repo).</p>
<p>If you have any interest in a career in data science/informatics, GitHub is also a common showcase of what (and how well/often) you can code. After some accumulation of code, definitely put your GitHub link on your CV/resume.</p>
<p>Check out the quickstart from github:
<a href="https://docs.github.com/en/get-started/quickstart/hello-world" class="uri">https://docs.github.com/en/get-started/quickstart/hello-world</a></p>
<h3 id="example-repos-rbi">Example repos (RBI)</h3>
<ul>
<li><a href="https://github.com/rnabioco/bmsc-7810-pbda">this class</a></li>
<li><a href="https://github.com/rnabioco/valr">valr</a></li>
<li><a href="https://github.com/wch/r-source">r-source</a> # mirror of R source code</li>
<li><a href="https://github.com/tidyverse/readr">readr</a></li>
</ul>
<h3 id="asking-for-help-with-other-packages-on-github">Asking for help with other packages on GitHub</h3>
<p>Every package should include README, installation instructions, and a maintained <code>issues</code> page where questions and bugs can be reported and addressed. Example: <a href="https://github.com/tidyverse/readr/issues">readr GitHub page</a> Donâ€™t be afraid to file new issues, but very often your problems are already answered in the <code>closed</code> section.</p>
<h2 id="finding-useful-packages">Finding useful packages</h2>
<p>In most cases, what you need is already made into well-documented packages, and you donâ€™t have to reinvent the wheel (but sometimes you should?). Depending on where the package is curated, installation is different. Some examples below:</p>
<ol type="1">
<li><code>Gviz</code> - visualize gene model</li>
<li><code>euler</code> - making custom euler/venn diagrams</li>
<li><code>emo</code> - inserting emojis into Rmd</li>
</ol>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># BiocManager::install("Gviz") # from bioconductor</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/utils/vignette.html'>vignette</a></span><span class='op'>(</span><span class='st'>"Gviz"</span><span class='op'>)</span></span>
<span></span>
<span><span class='co'># install.packages("eulerr") # from CRAN</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='fu'>eulerr</span><span class='fu'>::</span><span class='fu'><a href='https://jolars.github.io/eulerr/reference/euler.html'>euler</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>set1 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"geneA"</span>, <span class='st'>"geneB"</span>, <span class='st'>"geneC"</span><span class='op'>)</span>, </span>
<span>                       set2 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"geneC"</span>, <span class='st'>"geneD"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># devtools::install_github("hadley/emo") # from github</span></span>
<span><span class='fu'>emo</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/emo/man/ji.html'>ji</a></span><span class='op'>(</span><span class='st'>"smile"</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>#&gt; ðŸ˜„</code></pre>
</div>
<h2 id="bioconductor">Bioconductor</h2>
<p><img src="https://bioconductor.org/images/logo_bioconductor.gif" /></p>
<p>2,000+ R packages dedicated to bioinformatics. Included a coherent framework of data structures (e.g.Â SummarizedExperiment) built by dedicated Core members. Also includs many annotation and experimental datasets built into R packages and objects (See AnnotationHub and ExperimentHub)</p>
<ul>
<li><a href="https://bioconductor.org/" class="uri">https://bioconductor.org/</a><br />
</li>
<li>Use <code>BiocManager::install()</code> to install these packages<br />
</li>
<li><a href="https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html">RNA-seq workflow</a><br />
</li>
<li><a href="https://bioconductor.org/books/release/OSCA/">Orchestrating single cell analysis</a></li>
</ul>
<h2 id="finding-help-online">Finding help online</h2>
<p><img src="https://imgs.xkcd.com/comics/wisdom_of_the_ancients.png" /></p>
<p>The <a href="https://community.rstudio.com/">R studio community forums</a> are a great resource for asking questions about tidyverse related packages.</p>
<p><a href="https://stackoverflow.com/">StackOverflow</a> provides user-contributed questions and answers on a variety of topics.</p>
<p>For help with bioconductor packages, visit the <a href="https://support.bioconductor.org/">Bioc support page</a></p>
<p>Find out if others are having similar issues by searching the issue on the package <a href="https://github.com/">GitHub</a> page.</p>
<h2 id="cheat-sheets">Cheat sheets</h2>
<p>Rstudio links to common ones here: <code>Help</code> -&gt; <code>Cheatsheets</code>. More are hosted online, such as for <a href="https://rstudio.com/wp-content/uploads/2016/09/RegExCheatsheet.pdf">regular expressions</a>.</p>
<p>Useful to keep your own stash too.</p>
<h2 id="offline-help">Offline help</h2>
<p>The RBI fellows hold standing office hours on Thursdays over zoom. We are happy to help out with coding and RNA/DNA-related informatics questions. Send us an email to schedule a time (<code>rbi.fellows@cuanschutz.edu</code>).</p>
<h2 id="sometimes-code-is-just-broken">Sometimes code is just broken</h2>
<p>No one writes perfect code. Developers often expect that there will be bugs in their code. If you suspect bugs or mishandled edge cases, go to the package GitHub and search the issues section to see if the problem has been reported or fixed. If not, submit an issue that describes the problem.</p>
<p>The <a href="https://reprex.tidyverse.org/">reprex package</a> makes it easy to
produce well-formatted reproducible examples that demonstrate the problem. Often developers will be thankful for your help with making their software better.</p>
<h2 id="additional-resources">Additional Resources</h2>
<h3 id="generaldata-science">General/Data science</h3>
<ul>
<li><a href="https://rafalab.github.io/dsbook/">Introduction to Data Science</a></li>
<li><a href="https://stat545.com/">Stats in R</a></li>
<li><a href="https://bookdown.org/rdpeng/rprogdatascience/">R programming for data science</a></li>
<li><a href="https://r4ds.had.co.nz/index.html">R for data science</a></li>
<li><a href="https://adv-r.hadley.nz/">Advanced R</a></li>
<li><a href="https://github.com/matloff/fasteR">fasteR base R tutorial</a></li>
</ul>
<h3 id="genomics">Genomics</h3>
<ul>
<li><a href="http://genomicsclass.github.io/book/">PH525x series - Biomedical Data Science</a></li>
<li><a href="https://vincebuffalo.com/book/">Bioinformatics Data Skills</a></li>
<li><a href="https://www.biostarhandbook.com/">Biostar Handbook</a></li>
<li><a href="https://datacarpentry.org/genomics-workshop/">Genomics Workshop</a></li>
</ul>
<h3 id="a-meta-list-of-r-resources">A meta-list of R resources</h3>
<ul>
<li><a href="https://github.com/iamericfletcher/awesome-r-learning-resources" class="uri">https://github.com/iamericfletcher/awesome-r-learning-resources</a></li>
</ul>
<h3 id="writing-high-performance-r-functions-with-r-c">Writing high-performance R functions with R + C++</h3>
<ul>
<li><a href="http://adv-r.had.co.nz/Rcpp.html" class="uri">http://adv-r.had.co.nz/Rcpp.html</a><br />
</li>
<li><a href="https://dirk.eddelbuettel.com/code/rcpp.html" class="uri">https://dirk.eddelbuettel.com/code/rcpp.html</a><br />
</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>


<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom"></div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
